{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##dump-phasing-from-BAM\n",
    "\n",
    "dumps read info including special phasing tag into a TSV file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%run ipy_setup.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_refmap(bamread):\n",
    "    bpos = bamread.get_aligned_pairs(matches_only=True)\n",
    "    positions = [i[0] for i in bpos]\n",
    "    refpos =  [i[1] for i in bpos] # positions in reference\n",
    "    refmap = dict(zip(refpos, positions))\n",
    "    return refmap\n",
    "\n",
    "def get_matched_bases_in_read(bamread, in_pos, refmap):\n",
    "    outseq = [bamread.seq[refmap[i]] if i in refmap else 'N' for i in in_pos]\n",
    "    outseq = ''.join(outseq)\n",
    "    return outseq\n",
    "\n",
    "def get_position_in_read(bamread, in_pos, refmap):\n",
    "    outpos = None\n",
    "    if in_pos in refmap:\n",
    "        outpos = refmap[in_pos]\n",
    "    return outpos\n",
    "\n",
    "def reverse_compl(seq):\n",
    "    translate = {'A':'T', \n",
    "                 'T':'A', \n",
    "                 'C':'G',\n",
    "                 'G':'C', \n",
    "                 'N':'N'}\n",
    "    return ''.join([translate[s] for s in seq])\n",
    "\n",
    "class BlockVariant:\n",
    "    def __init__ (self, variantline):\n",
    "        # variant_id haplotype_1 haplotype_2 chromosome position refallele variantallele genotype allele_counts:genotype_likelihoods:delta:MEC_variant\n",
    "        ll = variantline.strip().split(\"\\t\")\n",
    "        var_id, hap1, hap2, chrom, pos, r_allele, v_allele, genotype, info_str = ll\n",
    "        self.chrom, self.r_allele, self.v_allele, self.info_str = chrom, r_allele, v_allele, info_str\n",
    "        self.var_id, self.hap1, self.hap2, self.pos = int(var_id), hap1, hap2, int(pos)\n",
    "        allele_counts, genotype_likelihoods, delta, MEC_variant = info_str.split(\":\")[0:4]\n",
    "        self.ref_count, self.alt_count = map(int, allele_counts.split(\",\"))\n",
    "        gen_00, gen_01, gen_11 = map(float, genotype_likelihoods.split(\",\"))\n",
    "        self.gen_like = {\"0/0\":gen_00, \"0/1\":gen_01, \"1/1\":gen_11}\n",
    "        self.delta = float(delta)\n",
    "        self.MEC_variant = MEC_variant\n",
    "    def __repr__ (self):\n",
    "        return \"<BlockVariant, var_id: %s>\" % str(self.var_id)\n",
    "\n",
    "\n",
    "class Block: # part of block reader\n",
    "    def __init__ (self, blockline):\n",
    "        # \"BLOCK: offset:\" first_variant_block \"len:\" length_of_block \"phased\": phased_variants_block SPAN: \n",
    "        # lengthspanned MECscore score fragments #fragments\n",
    "\n",
    "        ll               = blockline.strip().split()\n",
    "        self.offset      = int(ll[2])\n",
    "        self.total_len   = int(ll[4])\n",
    "        self.phased      = int(ll[6])\n",
    "        self.span        = int(ll[8])\n",
    "        self.MECscore    = float(ll[10])\n",
    "        self.fragments   = int(ll[12])\n",
    "        self.variants \t = dict() # default to empty\n",
    "        self.variant_ids = set()\n",
    "        self.chrom = None\n",
    "        self.start = None\n",
    "        self.end = None\n",
    "        self.informative_reads = []\n",
    "        self.unphased_reads = []\n",
    "        self.unphased_read_set = set()\n",
    "        self.read_count = 0\n",
    "        self.read_set = set()\n",
    "\n",
    "    def __repr__ (self):\n",
    "        return \"<Block, offset_id: %s>\" % str(self.offset)\n",
    "\n",
    "    def addVariant(self, variantline):\n",
    "        variant = BlockVariant(variantline)\n",
    "        self.variants[variant.var_id] = variant\n",
    "        self.variant_ids.add(variant.var_id)\n",
    "        self.updatePosition(variant)\n",
    "\n",
    "    def updatePosition(self, variant): # we need to do this because sometimes the variant isn't associated with a block\n",
    "        if (self.chrom != None) & (self.start != None) & (self.end != None):\n",
    "            if variant.pos < self.start:\n",
    "                self.start = variant.pos\n",
    "            elif variant.pos > self.end:\n",
    "                self.end = variant.pos\n",
    "            if self.chrom != variant.chrom:\n",
    "                sys.stderr.write(\"WARNING: Cannot add variants from other contigs to current block.\")\n",
    "                return 1\n",
    "        else:\n",
    "            self.chrom = variant.chrom\n",
    "            self.start = variant.pos\n",
    "            self.end = variant.pos\n",
    "\n",
    "    def concordance(self, input_read):\n",
    "        ''' this should return a dict of (#T,#F) tuples per variant\n",
    "         each element is a variant's concordance with the reads\n",
    "         using the read's haplotype information, we can establish whether the read's phasing\n",
    "         is consistent with how the variant was phased '''\n",
    "        variant_concord = dict()\n",
    "        support_reads_hap2 = 0\n",
    "        against_reads_hap2 = 0\n",
    "        support_reads_hap1 = 0\n",
    "        against_reads_hap1 = 0\n",
    "        for k,variant in self.variants.iteritems():\n",
    "            for read in input_reads:\n",
    "                if variant.var_id in read.positions:\n",
    "                    read_allele = read.alleles[read.positions.index(variant.var_id)]\n",
    "                    hapstate = read.haplotypes[self.offset]\n",
    "                    if hapstate == 2:\n",
    "                        if variant.hap2 != read_allele:\n",
    "                            against_reads_hap2 += 1\n",
    "                        else:\n",
    "                            support_reads_hap2 += 1\n",
    "                    else:\n",
    "                        if variant.hap1 != read_allele:\n",
    "                            against_reads_hap1 += 1\n",
    "                        else:\n",
    "                            support_reads_hap1 += 1\n",
    "        variant_concord[self.offset] = {\"hap1\": (support_reads_hap1, against_reads_hap1), \n",
    "                                               \"hap2\": (support_reads_hap2, against_reads_hap2)}\n",
    "        return variant_concord\n",
    "\n",
    "    def variant(self, var_id):\n",
    "        try:\n",
    "            return self.variants[var_id]\n",
    "        except:\n",
    "            return None\n",
    "\n",
    "class HapCutReader: # hapcut file reader\n",
    "\n",
    "    def __init__ ( self, fn ):\n",
    "        self.fn = fn\n",
    "        self.blocks = dict()\n",
    "        self.translate = dict()\n",
    "        for block in self.read_file_to_blocks(fn):\n",
    "            self.blocks[block.offset] = block\n",
    "            for v in block.variant_ids: # v is an id\n",
    "                self.translate[v] = block.offset\n",
    "\n",
    "    def loc(self, block_id):\n",
    "        try:\n",
    "            return self.blocks[self.translate[block_id]]\n",
    "        except:\n",
    "            return None\n",
    "\n",
    "    def read_file_to_blocks(self, fn):\n",
    "        with open(fn) as f:\n",
    "            currBlock = None\n",
    "            for l in f:\n",
    "                if l[0] == \"B\": # starting a new block\n",
    "                    currBlock = Block(l)\n",
    "                elif l[0] == \"*\": # ending a block\n",
    "                    yield currBlock\n",
    "                else:\n",
    "                    currBlock.addVariant(l)\n",
    "\n",
    "    def __repr__ (self):\n",
    "        return \"<HapCutReader, filename: %s>\" % self.fn\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[****************100%******************]  292 of 293 complete Elapsed: 91.5 minutes\t Remaining: 0.3 minutes\n"
     ]
    }
   ],
   "source": [
    "basedir = \"/hpc/users/neffr01/jason_new/hapcut_outputs/hg002_re_000000F\"\n",
    "bam_file = basedir + \"/hapcut_qv13_mq10/hg002.000000F.qv13mq10.redo.merged.ann.bam\"\n",
    "hapcut_file = basedir + \"/hapcut_qv13_mq10/hg002_hapcut_000000F.hapcut\"\n",
    "out_file = open(\"/hpc/users/neffr01/2work/hg002_dump_read_phasing.txt\", 'w')\n",
    "\n",
    "bam_reader = pysam.AlignmentFile(bam_file)\n",
    "block_reader = HapCutReader(hapcut_file)\n",
    "pbar = ProgressBar(len(block_reader.blocks))\n",
    "for bid, block in block_reader.blocks.iteritems():\n",
    "    pbar.animate()\n",
    "    block_variants = [(v.pos, v.r_allele, v.v_allele, int(v.hap1), int(v.hap2)) for v \\\n",
    "              in block.variants.values()]\n",
    "    for read in bam_reader.fetch(region=\"000000F\", start=block.start, end=block.end):\n",
    "        start, end, read_haps = read.reference_start, read.reference_end, []\n",
    "        refmap = get_refmap(read)\n",
    "        block_count = 0\n",
    "        try:\n",
    "            zhtag = read.get_tag(\"ZH\").split(\";\")\n",
    "            read_blocks = [int(a.split(\",\")[0]) for a in zhtag]\n",
    "            read_haps   = [int(a.split(\",\")[1]) for a in zhtag]\n",
    "            block_index = read_blocks.index(bid)\n",
    "        except:\n",
    "            continue # \n",
    "        block_hap   = read_haps[block_index]\n",
    "        read_variants = [x for x in block_variants if ((x[0] >= start) & (x[0] <= end))]\n",
    "        supp = {0:0, 1:0}\n",
    "        block_hap = block_hap-1\n",
    "        for var in read_variants:\n",
    "            readbase = get_matched_bases_in_read(read, [var[0]-1], refmap)\n",
    "            base_to_hap = { var[1]:var[3], \n",
    "                            var[2]:var[4]}\n",
    "            try:\n",
    "                supp[base_to_hap[readbase]] +=1\n",
    "            except:\n",
    "                continue\n",
    "        print >>out_file, bid, block_hap, supp[0], supp[1], read.qname + \"#\" + str(block_count)\n",
    "out_file.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
