{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "%run ./ipy_setup.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#!/usr/bin/env python\n",
    "\n",
    "# imports\n",
    "import os, sys, getopt\n",
    "import pysam\n",
    "from itertools import groupby\n",
    "from operator import itemgetter\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "global count\n",
    "count = 0\n",
    "\n",
    "# init main\n",
    "def main(argv):\n",
    "    hairs_file = ''\n",
    "    hapcut_file = ''\n",
    "    bam_file = ''\n",
    "    out_file = ''\n",
    "    help = 'greedy_partitioner.py -h <input.hairs> -c <input.hapcut> -i <input.bam> -o <output.ann.bam>'\n",
    "    try:\n",
    "        opts, args = getopt.getopt(argv,\"h:c:i:o:\",[\"hairs=\",\"hapcut=\", \"input=\", \"output=\"])\n",
    "    except getopt.GetoptError:\n",
    "        print help\n",
    "        sys.exit(2)\n",
    "    for opt, arg in opts:\n",
    "        if opt == '--help':\n",
    "            print help\n",
    "            sys.exit()\n",
    "        elif opt in (\"-h\", \"--hairs\"):\n",
    "            hairs_file = arg\n",
    "        elif opt in (\"-c\", \"--hapcut\"):\n",
    "            hapcut_file = arg\n",
    "        elif opt in (\"-i\", \"--input\"):\n",
    "            bam_file = arg\n",
    "        elif opt in (\"-o\", \"--output\"):\n",
    "            out_file = arg\n",
    "        else:\n",
    "            assert False, \"unhandled option\"\n",
    "\n",
    "    assert pysam.Samfile(bam_file, 'rb'), 'ERROR: Cannot open bam file for reading.'\n",
    "    assert open(bam_file + '.bai', 'rb'), 'ERROR: bam file is not indexed!'\n",
    "    bam_fp = pysam.Samfile(bam_file, 'rb')\n",
    "\n",
    "    if out_file==None:\n",
    "        out_file = bam_file + \".ann_haplotypes_\" + time.strftime(\"%m%d%y_%H%M%S\") + '.bam'\n",
    "\n",
    "    assert pysam.AlignmentFile(out_file, \"wb\", template=bam_fp), 'ERROR: Cannot open output file for writing.'\n",
    "    out_fp = pysam.AlignmentFile(out_file, \"wb\", template=bam_fp)\n",
    "\n",
    "    assert open(hairs_file), 'ERROR: Cannot access hairs file.'\n",
    "    assert open(hapcut_file), 'ERROR: Cannot open hapcut file.'\n",
    "\n",
    "    hair_reader = HairReader(hairs_file)\n",
    "    block_reader = HapCutReader(hapcut_file)\n",
    "    stats_file = out_file + \".interblock_stats.tsv\"\n",
    "    sys.stdout.write(\"Loaded greedy_partitoner.py, beginning execution. \\n\")\n",
    "    sys.stdout.write(\"Placing hairfile reads into haplotype blocks. \\n\")\n",
    "    sys.stdout.flush()\n",
    "    hair_reader.assemble_blocks(block_reader)\n",
    "    sys.stdout.write(\"Placing bamfile reads into haplotype blocks. \\n\")\n",
    "    sys.stdout.flush()\n",
    "    unphased_reads = unphased_heuristics(hair_reader, block_reader, bam_fp)\n",
    "    sys.stdout.write(\"Partitioning reads. \\n\")\n",
    "    sys.stdout.flush()\n",
    "    tag_reads(bam_fp, hair_reader, block_reader, unphased_reads, out_fp) #begin tagging reads\n",
    "    sys.stdout.write(\"Calculating interblock statistics. \\n\")\n",
    "    sys.stdout.flush()\n",
    "    interblock_stats(hair_reader, block_reader, stats_file, bam_fp) #generate interblock stats\n",
    "    bam_fp.close()\n",
    "    out_fp.close()\n",
    "\n",
    "# end of main\n",
    "\n",
    "### CLASSES ###\n",
    "\n",
    "class BlockVariant:\n",
    "    def __init__ (self, variantline):\n",
    "        self.original_line = variantline\n",
    "        # variant_id haplotype_1 haplotype_2 chromosome position refallele variantallele genotype allele_counts:genotype_likelihoods:delta:MEC_variant\n",
    "        ll = variantline.strip().split(\"\\t\")\n",
    "        var_id, hap1, hap2, chrom, pos, r_allele, v_allele, genotype, info_str = ll\n",
    "        self.chrom, self.r_allele, self.v_allele, self.info_str = chrom, r_allele, v_allele, info_str\n",
    "        self.var_id, self.hap1, self.hap2, self.pos = int(var_id), hap1, hap2, int(pos)\n",
    "        allele_counts, genotype_likelihoods, delta, MEC_variant = info_str.split(\":\")[0:4]\n",
    "        self.ref_count, self.alt_count = map(int, allele_counts.split(\",\"))\n",
    "        gen_00, gen_01, gen_11 = map(float, genotype_likelihoods.split(\",\"))\n",
    "        self.gen_like = {\"0/0\":gen_00, \"0/1\":gen_01, \"1/1\":gen_11}\n",
    "        self.delta = float(delta)\n",
    "        self.MEC_variant = MEC_variant\n",
    "    def __repr__ (self):\n",
    "        return \"<BlockVariant, var_id: %s>\" % str(self.var_id)\n",
    "\n",
    "\n",
    "class Block: # part of block reader\n",
    "    def __init__ (self, blockline):\n",
    "        # \"BLOCK: offset:\" first_variant_block \"len:\" length_of_block \"phased\": phased_variants_block SPAN: \n",
    "        # lengthspanned MECscore score fragments #fragments\n",
    "\n",
    "        ll               = blockline.strip().split()\n",
    "        self.offset      = int(ll[2])\n",
    "        self.total_len   = int(ll[4])\n",
    "        self.phased      = int(ll[6])\n",
    "        self.span        = int(ll[8])\n",
    "        self.MECscore    = float(ll[10])\n",
    "        self.fragments   = int(ll[12])\n",
    "        self.variants \t = dict() # default to empty\n",
    "        self.variant_ids = set()\n",
    "        self.chrom = None\n",
    "        self.start = None\n",
    "        self.end = None\n",
    "        self.informative_reads = []\n",
    "        self.unphased_reads = []\n",
    "        self.unphased_read_set = set()\n",
    "        self.read_count = 0\n",
    "        self.read_set = set()\n",
    "\n",
    "    def __repr__ (self):\n",
    "        return \"<Block, offset_id: %s>\" % str(self.offset)\n",
    "\n",
    "    def addVariant(self, variantline):\n",
    "        variant = BlockVariant(variantline)\n",
    "        self.variants[variant.var_id] = variant\n",
    "        self.variant_ids.add(variant.var_id)\n",
    "        self.updatePosition(variant)\n",
    "\n",
    "    def updatePosition(self, variant): # we need to do this because sometimes the variant isn't associated with a block\n",
    "        if (self.chrom != None) & (self.start != None) & (self.end != None):\n",
    "            if variant.pos < self.start:\n",
    "                self.start = variant.pos\n",
    "            elif variant.pos > self.end:\n",
    "                self.end = variant.pos\n",
    "            if self.chrom != variant.chrom:\n",
    "                sys.stderr.write(\"WARNING: Cannot add variants from other contigs to current block.\")\n",
    "                return 1\n",
    "        else:\n",
    "            self.chrom = variant.chrom\n",
    "            self.start = variant.pos\n",
    "            self.end = variant.pos\n",
    "\n",
    "    def addReadsToBlock(self, read_dict):\n",
    "        self.informative_reads = []\n",
    "        for k,read in read_dict.iteritems():\n",
    "            read_ids = frozenset([var[0] for block in read.blocks for var in block])\n",
    "            if len(set(read_ids).intersection(set(self.variant_ids))) > 0:\n",
    "                self.informative_reads.append(read)\n",
    "        self.read_count = len(self.informative_reads)\n",
    "        self.read_set = frozenset([x.read_id for x in self.informative_reads])\n",
    "\n",
    "    def concordance(self, input_reads):\n",
    "        ''' this should return a dict of (#T,#F) tuples per variant\n",
    "         each element is a variant's concordance with the reads\n",
    "         using the read's haplotype information, we can establish whether the read's phasing\n",
    "         is consistent with how the variant was phased '''\n",
    "        variant_concord = dict()\n",
    "        support_reads_hap2 = 0\n",
    "        against_reads_hap2 = 0\n",
    "        support_reads_hap1 = 0\n",
    "        against_reads_hap1 = 0\n",
    "        for k,variant in self.variants.iteritems():\n",
    "            for read in input_reads:\n",
    "                if variant.var_id in read.positions:\n",
    "                    read_allele = read.alleles[read.positions.index(variant.var_id)]\n",
    "                    hapstate = read.haplotypes[self.offset]\n",
    "                    if hapstate == 2:\n",
    "                        if variant.hap2 != read_allele:\n",
    "                            against_reads_hap2 += 1\n",
    "                        else:\n",
    "                            support_reads_hap2 += 1\n",
    "                    else:\n",
    "                        if variant.hap1 != read_allele:\n",
    "                            against_reads_hap1 += 1\n",
    "                        else:\n",
    "                            support_reads_hap1 += 1\n",
    "        variant_concord[self.offset] = {\"hap1\": (support_reads_hap1, against_reads_hap1), \n",
    "                                               \"hap2\": (support_reads_hap2, against_reads_hap2)}\n",
    "        return variant_concord\n",
    "\n",
    "    def variant(self, var_id):\n",
    "        try:\n",
    "            return self.variants[var_id]\n",
    "        except:\n",
    "            return None\n",
    "\n",
    "    def interblock_reads(self, input_reads):\n",
    "        out_reads = []\n",
    "        for read in input_reads:\n",
    "            if read.read_id not in self.read_set:\n",
    "                if read.chrom == self.chrom:\n",
    "                    if ((read.start < self.end) & (read.end > self.end)) | \\\n",
    "                        ((read.end > self.start) & (read.start < self.start)) | \\\n",
    "                        ((read.end <= self.end) & (read.start >= self.start)):\n",
    "                        out_reads.append(read)\n",
    "        return out_reads\n",
    "    \n",
    "    def add_read(self, read_obj):\n",
    "        if read_obj.read_id not in self.read_set:\n",
    "            self.informative_reads.append(read_obj)\n",
    "            self.read_set.add(read_obj.read_id)\n",
    "    \n",
    "    def add_unphased_reads(self, bam_fp):\n",
    "        reads = bam_fp.fetch(region=self.chrom + ':' + str(self.start) + '-' + str(self.end))\n",
    "        for read in reads:\n",
    "            if read.query_name not in self.read_set:\n",
    "                self.unphased_reads.append(read)\n",
    "                self.unphased_read_set.add(read.query_name)\n",
    "        return None\n",
    "\n",
    "class HapCutReader: # hapcut file reader\n",
    "\n",
    "    def __init__ ( self, fn ):\n",
    "        self.fn = fn\n",
    "        self.blocks = dict()\n",
    "        self.translate = dict()\n",
    "        for block in self.read_file_to_blocks(fn):\n",
    "            self.blocks[block.offset] = block\n",
    "            for v in block.variant_ids: # v is an id\n",
    "                self.translate[v] = block.offset\n",
    "\n",
    "    def loc(self, block_id):\n",
    "        try:\n",
    "            return self.blocks[self.translate[block_id]]\n",
    "        except:\n",
    "            return None\n",
    "\n",
    "    def read_file_to_blocks(self, fn):\n",
    "        with open(fn) as f:\n",
    "            currBlock = None\n",
    "            for l in f:\n",
    "                if l[0] == \"B\": # starting a new block\n",
    "                    currBlock = Block(l)\n",
    "                elif l[0] == \"*\": # ending a block\n",
    "                    yield currBlock\n",
    "                else:\n",
    "                    currBlock.addVariant(l)\n",
    "\n",
    "    def __repr__ (self):\n",
    "        return \"<HapCutReader, filename: %s>\" % self.fn\n",
    "    \n",
    "    def assemble_reads(self, hair_reader):\n",
    "        for k,block in self.blocks.iteritems():\n",
    "            block.addReadsToBlock(hair_reader.reads)\n",
    "\n",
    "class HapCutRead: #hair file line\n",
    "\n",
    "    def __init__ (self, hairline):\n",
    "        #Column 1 is the number of consecutive set of SNPs covered by the fragment, NOT haplotype blocks.\n",
    "        #Column 2 is the fragment id. \n",
    "        #Column 3 is the offset of the first block of SNPs covered by the fragment followed by the alleles at the SNPs in this block.\n",
    "        #Column 5 is the offset of the second block of SNPs covered by the fragment followed by the alleles at the SNPs in this block.\n",
    "        #...\n",
    "        #The last column is a string with the quality values (Sanger fastq format) for all the alleles covered by the fragment (concatenated for all blocks). \n",
    "        #For example, if a read/fragment covers SNPs 2,3 and 5 with the alleles 0, 1 and 0 respectively, then the input will be:\n",
    "        #2 read_id 2 01 5 0 AAC\n",
    "        #Here AAC is the string corresponding to the quality values at the three alleles. The encoding of 0/1 is arbitrary but following the VCF format, 0 is reference and 1 is alternate. \n",
    "        hairlist = hairline.strip().split()\n",
    "        self.blockcount = 0                # this information must be determined afterwards \n",
    "        self.read_id    = hairlist[1]      # read_id\n",
    "        self.blocks     = []\t\t       # an array of tuples corresponding to blocks\n",
    "        self.positions  = []\n",
    "        self.alleles    = []\n",
    "        self.chrom      = None\n",
    "        self.start      = None\n",
    "        self.end        = None\n",
    "        self.haplotypes = dict()             # an array of {\"block_offset\":\"haplotype\"} \n",
    "                                           # after partitioning\n",
    "        self.original_line = hairline\n",
    "        \n",
    "        for i in range(2, len(hairlist)-1, 2):\n",
    "            position = int(hairlist[i])\n",
    "            allele = hairlist[i+1]\n",
    "            block = zip(range(position, position+len(allele)), allele)\n",
    "            self.blocks.append(block)\n",
    "            self.positions.extend(range(position, position+len(allele)))\n",
    "            self.alleles.extend(allele)\n",
    "            self.qualities  = hairlist[-1]         # a matched arary of the qualities of allele calls\n",
    "\n",
    "    def __repr__(self):\n",
    "        return \"<HapCutRead, read_id: %s>\" % str(self.read_id)\n",
    "\n",
    "    def haplotype_fields(self):\n",
    "        haps = \";\".join([','.join([str(key), str(self.haplotypes[key])]) for key in self.haplotypes])\n",
    "        haptag = [(\"ZH\", haps), (\"ZB\", int(self.blockcount)), (\"ZV\", len(self.positions))]\n",
    "        return haptag\n",
    "    \n",
    "    def addGenomicPositions(self, block_reader):\n",
    "        arr = []\n",
    "        chrom = None\n",
    "        for position in self.positions:\n",
    "            b = block_reader.loc(position)\n",
    "            if b == None:\n",
    "                continue\n",
    "            if chrom == None:\n",
    "                chrom = b.chrom\n",
    "            arr.append(b.variant(position).pos)\n",
    "        if len(arr) > 0:\n",
    "            self.chrom = chrom\n",
    "            self.start = np.min(arr)\n",
    "            self.end = np.max(arr)\n",
    "        else:\n",
    "            self.chrom = '*'\n",
    "            self.start = None\n",
    "            self.end = None\n",
    "    \n",
    "    def assemble_blocks(self, block_reader):\n",
    "        # it turns out that the blocks provided in a hapcut file don't actually correspond to real blocks\n",
    "        # just contiguous alleles?\n",
    "        self.blocks = []\n",
    "        lastBlock = -1\n",
    "        for ix, pos in enumerate(self.positions): # self.positions corresponds to a variant id\n",
    "            currBlock = block_reader.loc(pos) # look up block associated with variant\n",
    "            if currBlock == None: # if it's not, continue\n",
    "                continue\n",
    "            # let's also add ourselves to the block\n",
    "            currBlock.add_read(self) #if we get a block back, add the read to the block's read set\n",
    "            currBlock = currBlock.offset # set our read's block id\n",
    "            if currBlock != lastBlock:\n",
    "                self.blocks.append([])\n",
    "            self.blocks[-1].append((pos, self.alleles[ix])) # a read can be more than one haplotype block long\n",
    "            lastBlock = currBlock\n",
    "        self.blockcount = len(self.blocks) # determine number of haplotype blocks read spans\n",
    "        self.addGenomicPositions(block_reader) # determine start-end positions of read\n",
    "    \n",
    "class HairReader:\n",
    "\n",
    "    def __init__ (self, fn):\n",
    "        self.fn = fn\n",
    "        self.reads = dict()\n",
    "        with open (fn) as f:\n",
    "            for l in f:\n",
    "                newread = HapCutRead(l)\n",
    "                self.reads[newread.read_id] = newread\n",
    "        self.read_set = frozenset(self.reads.keys())\n",
    "\n",
    "    def __repr__ (self):\n",
    "        return \"<HairReader, filename: %s>\" % self.fn\n",
    "\n",
    "    def loc(self, read_id):\n",
    "        try:\n",
    "            return self.reads[read_id]\n",
    "        except:\n",
    "            return None\n",
    "    \n",
    "    def assemble_blocks(self, block_reader):\n",
    "        for k,read in self.reads.iteritems():\n",
    "            read.assemble_blocks(block_reader)\n",
    "\n",
    "### FUNCTIONS ###\n",
    "\n",
    "'''\n",
    "\n",
    "tag_reads()\n",
    "\n",
    "Usage: Tags reads from a bam file corresponding to a particular haplotype, with haplotype\n",
    "definitions from HapCut, under the optional tag \"ZH\".\n",
    "\n",
    "Inputs:\n",
    "    bam_fp\n",
    "    A pysam.Samfile object pointing to the input file\n",
    "    hair_reader\n",
    "    A HairReader object pointing to the hairs file.\n",
    "    block_reader\n",
    "    A HapcutReader object pointing to the hapcut file.\n",
    "    out_fp\n",
    "    A pysam.AlignmentFile pointing to the output bam.\n",
    "Outputs:\n",
    "    (none - writes to out_fp)\n",
    "\n",
    "'''\n",
    "\n",
    "def tag_reads(bam_fp, hair_reader, block_reader, unphased_reads, out_fp):\n",
    "    ''' tag_reads(bam_fp, hair_reader, block_reader, out_fp)'''\n",
    "    global count\n",
    "    for bamread in bam_fp.fetch():\n",
    "        count += 1\n",
    "        if (count % 100) == 0:\n",
    "            sys.stdout.write(\"\\rWritten %s lines to output.\" % str(count))\n",
    "            sys.stdout.flush()\n",
    "        if bamread.query_name in hair_reader.read_set:\n",
    "            read = hair_reader.loc(bamread.query_name)\n",
    "            read = greedy_partition(read, block_reader)\n",
    "            bamread.tags += read.haplotype_fields()      # add the haplotype information to hairfile read\n",
    "            out_fp.write(bamread)\n",
    "        elif bamread.query_name in unphased_reads:\n",
    "            read = unphased_reads[bamread.query_name]\n",
    "            read = greedy_partition(read, block_reader)\n",
    "            bamread.tags += read.haplotype_fields()    # add the haplotype information to bam-based read\n",
    "            out_fp.write(bamread)\n",
    "        #out_fp.write(bamread) # write read to file\n",
    "\n",
    "'''\n",
    "greedy_partition()\n",
    "Ryan Neff\n",
    "\n",
    "inputs:\n",
    "read\n",
    "    a HapCutRead object\n",
    "block_reader\n",
    "    of the type HapCutReader\n",
    "\n",
    "outputs:\n",
    "    the original read, now with haplotype information.\n",
    "\n",
    "translate hairfile alleles into blockvar IDs\n",
    "get alleles in each read spanning blockvars\n",
    "determine alleles for the two blocks from blockvar\n",
    "partition read based on locally most probable alignment\n",
    "\n",
    "'''\n",
    "\n",
    "def greedy_partition(read, block_reader):\n",
    "    for readblock in read.blocks:\n",
    "        positions = [x[0] for x in readblock]\n",
    "        alleles = [x[1] for x in readblock]\n",
    "        allele_state = []\n",
    "        offset = positions[0]\n",
    "        hap = 0\n",
    "        block = block_reader.loc(offset) #retrieve block the read is in\n",
    "        if block == None: # this happens when hapcut throws out the block the read is in \n",
    "            continue\n",
    "        offset = block.offset\n",
    "        for ix, varpos in enumerate(positions):\n",
    "            blockvar = block.variant(varpos) # retrieve variant from VCF file\n",
    "            if blockvar.hap1 == alleles[ix]:\n",
    "                allele_state.append(-1)\n",
    "            elif blockvar.hap2 == alleles[ix]:\n",
    "                allele_state.append(1)\n",
    "            else:\n",
    "                #sys.stderr.write(\"\\nWarning: read allele matched no haplotypes.\")\n",
    "                #sys.stderr.write(\"\\nHair read: %s\" % read.read_id)\n",
    "                #sys.stderr.write(\"\\nAlleles: %s\" % str(alleles[ix]))\n",
    "                #sys.stderr.write(\"\\n Hap 1: %s, Hap 2: %s\\n\" % (blockvar.hap1, blockvar.hap2))\n",
    "                #sys.stderr.flush()\n",
    "                continue\n",
    "        if len(allele_state) < 1:\n",
    "            sys.stderr.write(\"Warning: no haplotype information in read.\\n\")\n",
    "            sys.stderr.write(\"\\nHair read: %s\" % read.read_id)\n",
    "            sys.stderr.write(\"\\nAlleles: %s\\n\" % str(alleles))\n",
    "            sys.stderr.flush()\n",
    "            hap = -1\n",
    "        if sum(allele_state) < 0:\n",
    "            hap = 1\n",
    "        elif sum(allele_state) > 0:\n",
    "            hap = 2\n",
    "        else:\n",
    "            hap = 0\n",
    "        read.haplotypes[offset] = hap\n",
    "    return read\n",
    "\n",
    "'''\n",
    "interblock_stats()\n",
    "\n",
    "Usage: Creates a tab-separated values file with statistics about reads overlapping\n",
    "between nearby blocks, and finds the concordance of these interblock reads\n",
    "with haplotypes in other blocks. \n",
    "\n",
    "inputs:\n",
    "    hair_reader\n",
    "        A HairReader object\n",
    "    block_reader\n",
    "        A HapcutReader object\n",
    "    out_stats\n",
    "        A string where the .tsv should be written. Defaults\n",
    "        to the hairs filename given in the input + 'interblock_stats.tsv'\n",
    "outputs:\n",
    "    none-writes to file directly\n",
    "\n",
    "'''\n",
    "\n",
    "def interblock_stats(hair_reader, block_reader, out_stats, bam_fp):\n",
    "    blockdist = []\n",
    "    lastChr = None\n",
    "    lastPos = None\n",
    "    lastBlock = None\n",
    "    lastReads = set()\n",
    "    for k,read in hair_reader.reads.iteritems():\n",
    "        if read.haplotypes == dict():\n",
    "            read = greedy_partition(read, block_reader)\n",
    "    for ix, key in enumerate(sorted(block_reader.blocks.keys())):\n",
    "        sys.stdout.write('\\r%s percent done.' % round(ix/float(len(block_reader.blocks))*100))\n",
    "        sys.stdout.flush()\n",
    "        block = block_reader.blocks[key]\n",
    "        if block.read_set == set():\n",
    "            block.addReadsToBlock(hair_reader.reads)\n",
    "        currBlock = block.offset\n",
    "        currChr = block.chrom\n",
    "        currPos = block.start\n",
    "        if lastBlock != None:\n",
    "            if lastChr == currChr:\n",
    "                interblock_reads = block.interblock_reads(lastBlock_obj.informative_reads)\n",
    "                row=[lastBlock, currBlock, currChr, lastPos, currPos, currPos-lastPos,\n",
    "                     lastBlock_obj.end-lastBlock_obj.start, block.end-block.start,\n",
    "                     len(lastBlock_obj.variant_ids), len(block.variant_ids), \n",
    "                     len(lastBlock_obj.read_set), len(block.read_set),\n",
    "                     len(list(bam_fp.fetch(region=lastChr + ':' + str(lastBlock_obj.start) + '-' + str(lastBlock_obj.end)))),\n",
    "                     len(list(bam_fp.fetch(region=lastChr + ':' + str(block.start) + '-' + str(block.end)))),\n",
    "                     len(interblock_reads),\n",
    "                     len(list(bam_fp.fetch(region=lastChr + ':' + str(lastBlock_obj.end) + '-' + str(block.start))))]\n",
    "                     #lastBlock_obj.concordance(lastBlock_obj.informative_reads), \n",
    "                     #block.concordance(block.informative_reads)]\n",
    "                blockdist.append(row)\n",
    "            else:\n",
    "                continue\n",
    "        lastBlock = currBlock\n",
    "        lastBlock_obj = block\n",
    "        lastChr = currChr\n",
    "        lastPos = block.end\n",
    "    header = ['block1', 'block2', 'chrom', 'block1_end', 'block2_start', \n",
    "              'distance', 'block1_size', 'block2_size', 'block1_variants', 'block2_variants', \n",
    "              'block1_informative_reads', 'block2_informative_reads', \n",
    "              'block1_reads', 'block2_reads',\n",
    "              'informative_interblock_reads', \n",
    "              'all_interblock_reads'] \n",
    "              #'block1_concordance', 'block2_concordance']\n",
    "    info = pd.DataFrame(blockdist, columns=header)\n",
    "    info.to_csv(out_stats, sep=\"\\t\")\n",
    "\n",
    "    # use freebayes call to modify the reference (reference mask to N)\n",
    "# look at reads that span more than two SNPs\n",
    "    # list(bam_fp.fetch(region=lastChr + ':' + str(block.start) + '-' + str(block.end)))\n",
    "    # see if the SNP was there or not - if it's there but the alignment is messed up we may need to modify it\n",
    "# compare haplotype calls to hg003 and hg004\n",
    "\n",
    "def reverse_compl(seq):\n",
    "    translate = {'A':'T', \n",
    "                 'T':'A', \n",
    "                 'C':'G',\n",
    "                 'G':'C'}\n",
    "    return ''.join([translate[s] for s in seq])\n",
    "\n",
    "def hamming_dist(str1, str2):\n",
    "    difference = 0\n",
    "    for x,y in zip(str1, str2):\n",
    "        if x != y:\n",
    "            difference += 1\n",
    "    return difference\n",
    "\n",
    "# get aligned portion per region\n",
    "# IPD data \n",
    "def get_matched_bases_in_read(bamread, in_pos):\n",
    "    bpos = bamread.get_aligned_pairs(matches_only=True)\n",
    "    positions = [i[0] for i in bpos]\n",
    "    refpos =  [i[1] for i in bpos] # positions in reference\n",
    "    refmap = dict(zip(refpos, positions))\n",
    "    outseq = [bamread.seq[refmap[i]] if i in refmap else 'N' for i in in_pos]\n",
    "    outseq = ''.join(outseq)\n",
    "    return outseq, ''\n",
    "\n",
    "# we can make this program phase the remaining reads\n",
    "def unphased_heuristics(hair_reader, block_reader, bam_fp):\n",
    "    #1. identify unphased reads per block\n",
    "    unphased_reads = dict()\n",
    "    uphase_read_set = set()\n",
    "    #pbar = ProgressBar(len(block_reader.blocks))\n",
    "    for bid, block in block_reader.blocks.iteritems():\n",
    "        block.add_unphased_reads(bam_fp)\n",
    "        pbar2 = ProgressBar(len(block.unphased_reads))\n",
    "        for bamread in block.unphased_reads:\n",
    "            pbar2.animate()\n",
    "            bs, be = bamread.reference_start, bamread.reference_end # 0-based indexing\n",
    "            br_variants = []\n",
    "            for variant in block.variants.values():\n",
    "                if (variant.pos >= bs) & (variant.pos <= be):\n",
    "                    len_var = len(variant.r_allele)\n",
    "                    var_id = variant.var_id\n",
    "                    refcall = variant.r_allele \n",
    "                    altcall = variant.v_allele.split(',') # array of possible alternate alleles\n",
    "                    read_bases, qual_vals = get_matched_bases_in_read(bamread,range(variant.pos-1, variant.pos-1+len_var))\n",
    "                    # now let's do some primitive variant calling\n",
    "                    read_call = None # this will be an integer corresponding to the variant call of the read\n",
    "                    read_dist = [hamming_dist(refcall, read_bases)]\n",
    "                    for a in altcall:\n",
    "                        read_dist.append(hamming_dist(a, read_bases))\n",
    "                    mindist = np.min(read_dist)\n",
    "                    min_allele = [i for i, x in enumerate(read_dist) if x == mindist]\n",
    "                    if len(min_allele) == 1:\n",
    "                        read_call = min_allele[0]\n",
    "                    # if we got a variant call, let's add it to the array\n",
    "                    if read_call != None:\n",
    "                        br_variants.append([var_id, read_call, qual_vals])\n",
    "            # end variant loop\n",
    "            if len(br_variants) > 0:\n",
    "                # let's print a hairreader line\n",
    "                ranges = []\n",
    "                qline = ''\n",
    "                for k, g in groupby(enumerate(br_variants), lambda (i,x):i-x[0]):\n",
    "                    group = map(itemgetter(1), g)\n",
    "                    ranges.append((group[0][0], group))\n",
    "                hairline = [str(len(ranges)), bamread.query_name]\n",
    "                for gid, group in ranges:\n",
    "                    hairline.append(str(gid))\n",
    "                    gline = ''\n",
    "                    for g in group:\n",
    "                        gline += str(g[1])\n",
    "                        qline += g[2]\n",
    "                    hairline.append(gline)\n",
    "                hairline.append(qline)\n",
    "                hairline = ' '.join(hairline)\n",
    "                '''this now exactly matches the hairs file format'''\n",
    "                # print hairline \n",
    "                if bamread.query_name not in uphase_read_set:\n",
    "                    hairread = HapCutRead(hairline)\n",
    "                    hairread.chrom = block.chrom\n",
    "                    hairread.start = bs\n",
    "                    hairread.end = be\n",
    "                    hairread.blockcount = 1\n",
    "                    hairread.blocks = [[(pos,str(allele)) for pos, allele, qual in br_variants]]\n",
    "                    unphased_reads[bamread.query_name] = hairread\n",
    "                    uphase_read_set.add(bamread.query_name)\n",
    "                else:\n",
    "                    other_read = unphased_reads[bamread.query_name]\n",
    "                    other_read.blockcount += 1\n",
    "                    other_read.blocks.append([(pos,str(allele)) for pos, allele, qual in br_variants])\n",
    "                    other_read.positions.extend([pos for pos, allele, qual in br_variants])\n",
    "                    other_read.alleles.extend([str(allele) for pos, allele, qual in br_variants])\n",
    "            # end check for variants in bamread\n",
    "            # break\n",
    "        # end bamread loop\n",
    "        # break\n",
    "    # end block loop\n",
    "    return unphased_reads\n",
    "\n",
    "# run the program if called from the command line\n",
    "#if __name__ == \"__main__\":\n",
    "#   main(sys.argv[1:])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def get_refmap(bamread):\n",
    "    bpos = bamread.get_aligned_pairs(matches_only=True)\n",
    "    positions = [i[0] for i in bpos]\n",
    "    refpos =  [i[1] for i in bpos] # positions in reference\n",
    "    refmap = dict(zip(refpos, positions))\n",
    "    return refmap\n",
    "\n",
    "def get_matched_bases_in_read(bamread, in_pos, refmap):\n",
    "    outseq = [bamread.seq[refmap[i]] if i in refmap else 'N' for i in in_pos]\n",
    "    outqual = [bamread.qual[refmap[i]] if i in refmap else '.' for i in in_pos]\n",
    "    outseq = ''.join(outseq)\n",
    "    outqual = ''.join(outqual)\n",
    "    return outseq, outqual\n",
    "\n",
    "def get_position_in_read(bamread, in_pos, refmap):\n",
    "    outpos = None\n",
    "    if in_pos in refmap:\n",
    "        outpos = refmap[in_pos]\n",
    "    return outpos\n",
    "\n",
    "def reverse_compl(seq):\n",
    "    translate = {'A':'T', \n",
    "                 'T':'A', \n",
    "                 'C':'G',\n",
    "                 'G':'C', \n",
    "                 'N':'N'}\n",
    "    return ''.join([translate[s] for s in seq])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# partitioning code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "basedir = '/sc/orga/scratch/bashia02/collaborations/hardik_shah/jason_new/hapcut_outputs/hg002_re_000000F/'\n",
    "bam_fp = pysam.Samfile(basedir + 'hg002_000000F.new.merged.bam', 'rb')\n",
    "#out_fp = pysam.AlignmentFile(basedir +'hapcut_qv13_mq10/hg002.000000F.qv13mq10.redo.merged.ann.bam', 'wb', template=bam_fp)\n",
    "hairs_file = basedir + 'hapcut_qv13_mq10/hg002_hapcut_000000F.hairs'\n",
    "hapcut_file = basedir + 'hapcut_qv13_mq10/hg002_hapcut_000000F.hapcut'\n",
    "\n",
    "hair_reader = HairReader(hairs_file)\n",
    "block_reader = HapCutReader(hapcut_file)\n",
    "hair_reader.assemble_blocks(block_reader)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#add methylation sites to hairs file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[*****************98%***************** ]  9765 of 10000 complete Elapsed: 1.7 minutes\t Remaining: 0.0 minutes\n"
     ]
    }
   ],
   "source": [
    "pbar = ProgressBar(10000)\n",
    "for i in range(1,10000):\n",
    "    pbar.animate()\n",
    "    time.sleep(0.01)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[*****************56%*                 ]  34507 of 61100 complete Elapsed: 0.0 minutes\t Remaining: 0.0 minutes\n"
     ]
    }
   ],
   "source": [
    "siteslist = []\n",
    "pbar = ProgressBar(61100)\n",
    "for line in open(\"/hpc/users/neffr01/jason_new/hapcut_outputs/hg002_qv13_methyl_sig_sites-novariant-11mer.tsv\", 'r'):\n",
    "    pbar.animate()\n",
    "    chrom, pos, mean1, mean2 = line.strip('\\n').split(' ')\n",
    "    pos, mean1, mean2 = int(pos), float(mean1), float(mean2)\n",
    "    siteslist.append([chrom, pos, mean1, mean2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sigoverlap = [i for i in siteslist if (i[1] <= rend) & (i[1] >= rstart)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "block_variants = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10240\t1\t0\t000000F\t13771002\tA\tG\t0/1:160.002:57:29:1076:28:1059:-94.7066,0,-96.2294\t2,1:-1.4,-0.1,-2.9:-1.2:0.0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for bid, block in block_reader.blocks.iteritems():\n",
    "    for vid, var in block.variants.iteritems():\n",
    "        print var.original_line\n",
    "        break\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from itertools import groupby, count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[*****************98%***************** ]  440135 of 450000 complete Elapsed: 565.3 minutes\t Remaining: 12.7 minutes\n"
     ]
    }
   ],
   "source": [
    "# going to need to regenerate hairs file from scratch because we lost information on the read\n",
    "out_new_hair = open(\"/hpc/users/neffr01/jason_new/methyl_predict/hg002_faked_hair.hairs\", 'w')\n",
    "pbar = ProgressBar(450000)\n",
    "counter = 0\n",
    "for read in bam_fp.fetch():\n",
    "    pbar.animate()\n",
    "    counter += 1\n",
    "    if counter % 100 == 0:\n",
    "        out_new_hair.flush()\n",
    "    refmap, ipd_values = None, None\n",
    "    start, end = read.reference_start, read.reference_end\n",
    "    read_variants = mergedvcf[(mergedvcf[1] >= start) & (mergedvcf[1] <= end)]\n",
    "    if len(read_variants) != 0:\n",
    "        refmap = get_refmap(read)\n",
    "        ipd_values = [int(i) for i in read.get_tag('ip').strip('S').split(',')]\n",
    "    phased_variants = []\n",
    "    quals = ''\n",
    "    for varid, varline in read_variants.iterrows():\n",
    "        varid += 1 # this gives us the correct line numbers\n",
    "        refpos = varline[1]-1\n",
    "        allele = 0\n",
    "        if 'methyl' in varline[2]:\n",
    "            # don't know if this is going to work... we may need to add some sort of cutoff here.\n",
    "            # but let's call the local methyl status and then hope the global optimum is concordant\n",
    "            basepos = get_position_in_read(read, refpos, refmap)\n",
    "            mean1 = float(varline[7].split(\";\")[0].split(\"=\")[1])\n",
    "            mean2 = float(varline[7].split(\";\")[1].split(\"=\")[1])\n",
    "            if mean1 > mean2: # swap them so the highest methyl value is always in the 1 position\n",
    "                tmp = mean2\n",
    "                mean2 = mean1\n",
    "                mean1 = tmp\n",
    "            if basepos==None: continue\n",
    "            ipd_value = np.log(ipd_values[basepos])\n",
    "            mean1_res = abs(mean1-ipd_value)\n",
    "            mean2_res = abs(mean2-ipd_value)\n",
    "            if abs(mean1_res-mean2_res) >= 0.5: # this is our calling cutoff (arbitrary for now...)\n",
    "                if mean1_res > mean2_res:\n",
    "                    allele = 1\n",
    "                phased_variants.append((varid, allele))\n",
    "                quals += '.'\n",
    "        else:\n",
    "            # let's get the base from the read and compare that to the vcf\n",
    "            # MAKE SURE that it is a het variant\n",
    "            try:\n",
    "                allelefreq = [float(a.split(\"=\")[1]) for a in varline[7].split(\";\") if 'AF' in a][0]\n",
    "            except:\n",
    "                continue # no multiallelic sites!\n",
    "            if allelefreq != 0.5:\n",
    "                continue\n",
    "            refallele = varline[3]\n",
    "            altallele = varline[4]\n",
    "            alen = len(refallele)\n",
    "            read_bases, read_qual = get_matched_bases_in_read(read,range(refpos, refpos+alen),refmap)\n",
    "            ref_dist = hamming_dist(refallele, read_bases)\n",
    "            alt_dist = hamming_dist(altallele, read_bases)\n",
    "            if ref_dist != alt_dist:\n",
    "                if ref_dist > alt_dist:\n",
    "                    allele = 1\n",
    "                phased_variants.append((varid, allele))\n",
    "                quals += read_qual[0] \n",
    "    groups = groupby(phased_variants, key=lambda item, c=count():item[0]-next(c))\n",
    "    tmp = [list(g) for k, g in groups]\n",
    "    blockcount = len(tmp)\n",
    "    if blockcount > 2: \n",
    "        continue\n",
    "    outstr = ''\n",
    "    outstr += str(blockcount) + \" \" + read.qname + \" \"\n",
    "    for site in tmp:\n",
    "        outstr += str(site[0][0]) + \" \"+''.join([str(y[1]) for y in site]) + \" \"\n",
    "    outstr += quals\n",
    "    print >>out_new_hair, outstr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5 m150129_030257_42163R_c100761962550000001823157107221505_s1_p0/124655/25653_36745 23177 1 23179 00 23182 0 23185 0 23187 0 .///..\n",
      "\n",
      "23177\n",
      "23179\n",
      "23180\n",
      "23182\n",
      "23185\n",
      "23187\n"
     ]
    }
   ],
   "source": [
    "for hid, hair in hair_reader.reads.iteritems():\n",
    "    print hair.original_line\n",
    "    for var in hair.positions:\n",
    "        print var\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "siteslist = []\n",
    "pbar = ProgressBar(61100)\n",
    "for line in open(\"/hpc/users/neffr01/jason_new/hapcut_outputs/hg002_re_000000F/ipds_fwd/hg002_fwd_all_positions_000000F:0-130118.tsv_sig\", 'r'):\n",
    "    pbar.animate()\n",
    "    chrom, pos, mean1, mean2 = line.strip('\\n').split('\\t', 5)[0:4]\n",
    "    pos, mean1, mean2 = int(pos), float(mean1), float(mean2)\n",
    "    if mean1 > mean2:\n",
    "        tmp = mean1\n",
    "        mean1 = mean2\n",
    "        mean2 = tmp\n",
    "    siteslist.append([chrom, pos, mean1, mean2])\n",
    "methylVCF = open(\"/hpc/users/neffr01/jason_new/hapcut_outputs/hg002_re_000000F/ipds_fwd/hg002_fwd_all_positions_000000F:0-130118.tsv_sig.vcf\", 'w')\n",
    "siteslist[0]\n",
    "count = 0\n",
    "for a in siteslist:\n",
    "    count += 1\n",
    "    print >>methylVCF, '\\t'.join([str(val) for val in [a[0], a[1], 'methyl-' + str(count), \\\n",
    "                                    'X', 'MX', 100.00, '.', \"MEAN1=\" + str(round(a[2],3)) + ';MEAN2=' \\\n",
    "                                    + str(round(a[3], 3)), 'GT', '0/1']])\n",
    "methylVCF.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "origvcf = pd.read_csv('/hpc/users/neffr01/jason_new/hapcut_outputs/hg002_re_000000F/hg002_short_reads_rg-rmdup_000000F.freebayes.vcf_snps.nohead.vcf', \n",
    "                      sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "methylvcf = pd.read_csv('/hpc/users/neffr01/jason_new/methyl_predict/methyl_vcf_hapcut_qv13_sig_sites.vcf', \n",
    "                      sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "origvcf.set_index(origvcf[1], inplace=True)\n",
    "methylvcf.set_index(methylvcf[1], inplace=True)\n",
    "\n",
    "mergedvcf = pd.concat([origvcf, methylvcf]).sort()\n",
    "mergedvcf.to_csv(\"/hpc/users/neffr01/jason_new/methyl_predict/hg002_shortread-vars_000000F_merged_methyl_sig.vcf\", \n",
    "                sep='\\t', header=None, index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>000000F</td>\n",
       "      <td>43</td>\n",
       "      <td>methyl-1</td>\n",
       "      <td>X</td>\n",
       "      <td>MX</td>\n",
       "      <td>100.00</td>\n",
       "      <td>.</td>\n",
       "      <td>MEAN1=2.373;MEAN2=3.243</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>000000F</td>\n",
       "      <td>125</td>\n",
       "      <td>methyl-2</td>\n",
       "      <td>X</td>\n",
       "      <td>MX</td>\n",
       "      <td>100.00</td>\n",
       "      <td>.</td>\n",
       "      <td>MEAN1=3.087;MEAN2=2.231</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>000000F</td>\n",
       "      <td>613</td>\n",
       "      <td>methyl-3</td>\n",
       "      <td>X</td>\n",
       "      <td>MX</td>\n",
       "      <td>100.00</td>\n",
       "      <td>.</td>\n",
       "      <td>MEAN1=3.655;MEAN2=2.698</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>000000F</td>\n",
       "      <td>647</td>\n",
       "      <td>methyl-4</td>\n",
       "      <td>X</td>\n",
       "      <td>MX</td>\n",
       "      <td>100.00</td>\n",
       "      <td>.</td>\n",
       "      <td>MEAN1=2.993;MEAN2=2.119</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>000000F</td>\n",
       "      <td>966</td>\n",
       "      <td>methyl-5</td>\n",
       "      <td>X</td>\n",
       "      <td>MX</td>\n",
       "      <td>100.00</td>\n",
       "      <td>.</td>\n",
       "      <td>MEAN1=3.628;MEAN2=2.617</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>000000F</td>\n",
       "      <td>1281</td>\n",
       "      <td>methyl-6</td>\n",
       "      <td>X</td>\n",
       "      <td>MX</td>\n",
       "      <td>100.00</td>\n",
       "      <td>.</td>\n",
       "      <td>MEAN1=3.102;MEAN2=2.553</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>000000F</td>\n",
       "      <td>1520</td>\n",
       "      <td>methyl-7</td>\n",
       "      <td>X</td>\n",
       "      <td>MX</td>\n",
       "      <td>100.00</td>\n",
       "      <td>.</td>\n",
       "      <td>MEAN1=2.954;MEAN2=2.018</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>000000F</td>\n",
       "      <td>1581</td>\n",
       "      <td>methyl-8</td>\n",
       "      <td>X</td>\n",
       "      <td>MX</td>\n",
       "      <td>100.00</td>\n",
       "      <td>.</td>\n",
       "      <td>MEAN1=3.041;MEAN2=1.927</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>000000F</td>\n",
       "      <td>1582</td>\n",
       "      <td>methyl-9</td>\n",
       "      <td>X</td>\n",
       "      <td>MX</td>\n",
       "      <td>100.00</td>\n",
       "      <td>.</td>\n",
       "      <td>MEAN1=3.193;MEAN2=2.476</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>000000F</td>\n",
       "      <td>1735</td>\n",
       "      <td>methyl-10</td>\n",
       "      <td>X</td>\n",
       "      <td>MX</td>\n",
       "      <td>100.00</td>\n",
       "      <td>.</td>\n",
       "      <td>MEAN1=2.553;MEAN2=3.254</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>000000F</td>\n",
       "      <td>1804</td>\n",
       "      <td>methyl-11</td>\n",
       "      <td>X</td>\n",
       "      <td>MX</td>\n",
       "      <td>100.00</td>\n",
       "      <td>.</td>\n",
       "      <td>MEAN1=3.662;MEAN2=4.263</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>000000F</td>\n",
       "      <td>1870</td>\n",
       "      <td>methyl-12</td>\n",
       "      <td>X</td>\n",
       "      <td>MX</td>\n",
       "      <td>100.00</td>\n",
       "      <td>.</td>\n",
       "      <td>MEAN1=3.015;MEAN2=2.489</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>000000F</td>\n",
       "      <td>2001</td>\n",
       "      <td>.</td>\n",
       "      <td>ATTA</td>\n",
       "      <td>CATT</td>\n",
       "      <td>1714.17</td>\n",
       "      <td>.</td>\n",
       "      <td>AB=0;ABP=0;AC=2;AF=1;AN=2;AO=53;CIGAR=2X1M1X;D...</td>\n",
       "      <td>GT:GQ:DP:RO:QR:AO:QA:GL</td>\n",
       "      <td>1/1:160.002:53:0:0:53:1948:-175.688,-15.9546,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>000000F</td>\n",
       "      <td>2328</td>\n",
       "      <td>methyl-13</td>\n",
       "      <td>X</td>\n",
       "      <td>MX</td>\n",
       "      <td>100.00</td>\n",
       "      <td>.</td>\n",
       "      <td>MEAN1=3.335;MEAN2=2.75</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>000000F</td>\n",
       "      <td>2363</td>\n",
       "      <td>methyl-14</td>\n",
       "      <td>X</td>\n",
       "      <td>MX</td>\n",
       "      <td>100.00</td>\n",
       "      <td>.</td>\n",
       "      <td>MEAN1=2.501;MEAN2=2.997</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>000000F</td>\n",
       "      <td>2387</td>\n",
       "      <td>methyl-15</td>\n",
       "      <td>X</td>\n",
       "      <td>MX</td>\n",
       "      <td>100.00</td>\n",
       "      <td>.</td>\n",
       "      <td>MEAN1=3.328;MEAN2=2.507</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>000000F</td>\n",
       "      <td>2675</td>\n",
       "      <td>methyl-16</td>\n",
       "      <td>X</td>\n",
       "      <td>MX</td>\n",
       "      <td>100.00</td>\n",
       "      <td>.</td>\n",
       "      <td>MEAN1=2.513;MEAN2=2.872</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>000000F</td>\n",
       "      <td>3089</td>\n",
       "      <td>methyl-17</td>\n",
       "      <td>X</td>\n",
       "      <td>MX</td>\n",
       "      <td>100.00</td>\n",
       "      <td>.</td>\n",
       "      <td>MEAN1=3.296;MEAN2=2.48</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>000000F</td>\n",
       "      <td>3638</td>\n",
       "      <td>methyl-18</td>\n",
       "      <td>X</td>\n",
       "      <td>MX</td>\n",
       "      <td>100.00</td>\n",
       "      <td>.</td>\n",
       "      <td>MEAN1=3.161;MEAN2=2.709</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>000000F</td>\n",
       "      <td>3741</td>\n",
       "      <td>methyl-19</td>\n",
       "      <td>X</td>\n",
       "      <td>MX</td>\n",
       "      <td>100.00</td>\n",
       "      <td>.</td>\n",
       "      <td>MEAN1=2.692;MEAN2=3.728</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>000000F</td>\n",
       "      <td>4284</td>\n",
       "      <td>methyl-20</td>\n",
       "      <td>X</td>\n",
       "      <td>MX</td>\n",
       "      <td>100.00</td>\n",
       "      <td>.</td>\n",
       "      <td>MEAN1=3.013;MEAN2=2.549</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>000000F</td>\n",
       "      <td>4371</td>\n",
       "      <td>methyl-21</td>\n",
       "      <td>X</td>\n",
       "      <td>MX</td>\n",
       "      <td>100.00</td>\n",
       "      <td>.</td>\n",
       "      <td>MEAN1=2.792;MEAN2=2.477</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>000000F</td>\n",
       "      <td>4645</td>\n",
       "      <td>methyl-22</td>\n",
       "      <td>X</td>\n",
       "      <td>MX</td>\n",
       "      <td>100.00</td>\n",
       "      <td>.</td>\n",
       "      <td>MEAN1=2.267;MEAN2=2.932</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>000000F</td>\n",
       "      <td>4648</td>\n",
       "      <td>methyl-23</td>\n",
       "      <td>X</td>\n",
       "      <td>MX</td>\n",
       "      <td>100.00</td>\n",
       "      <td>.</td>\n",
       "      <td>MEAN1=2.876;MEAN2=3.478</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>000000F</td>\n",
       "      <td>5164</td>\n",
       "      <td>methyl-24</td>\n",
       "      <td>X</td>\n",
       "      <td>MX</td>\n",
       "      <td>100.00</td>\n",
       "      <td>.</td>\n",
       "      <td>MEAN1=2.692;MEAN2=3.056</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>000000F</td>\n",
       "      <td>5351</td>\n",
       "      <td>methyl-25</td>\n",
       "      <td>X</td>\n",
       "      <td>MX</td>\n",
       "      <td>100.00</td>\n",
       "      <td>.</td>\n",
       "      <td>MEAN1=3.278;MEAN2=2.582</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>000000F</td>\n",
       "      <td>5369</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>805.85</td>\n",
       "      <td>.</td>\n",
       "      <td>AB=0.530612;ABP=3.40914;AC=1;AF=0.5;AN=2;AO=26...</td>\n",
       "      <td>GT:GQ:DP:RO:QR:AO:QA:GL</td>\n",
       "      <td>0/1:160.002:49:23:839:26:969:-86.5982,0,-74.8903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>000000F</td>\n",
       "      <td>6153</td>\n",
       "      <td>methyl-26</td>\n",
       "      <td>X</td>\n",
       "      <td>MX</td>\n",
       "      <td>100.00</td>\n",
       "      <td>.</td>\n",
       "      <td>MEAN1=3.199;MEAN2=2.701</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>000000F</td>\n",
       "      <td>6259</td>\n",
       "      <td>methyl-27</td>\n",
       "      <td>X</td>\n",
       "      <td>MX</td>\n",
       "      <td>100.00</td>\n",
       "      <td>.</td>\n",
       "      <td>MEAN1=3.066;MEAN2=2.208</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>000000F</td>\n",
       "      <td>6310</td>\n",
       "      <td>methyl-28</td>\n",
       "      <td>X</td>\n",
       "      <td>MX</td>\n",
       "      <td>100.00</td>\n",
       "      <td>.</td>\n",
       "      <td>MEAN1=2.467;MEAN2=1.85</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114353</th>\n",
       "      <td>000000F</td>\n",
       "      <td>33303233</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>754.60</td>\n",
       "      <td>.</td>\n",
       "      <td>AB=0.471698;ABP=3.37904;AC=1;AF=0.5;AN=2;AO=25...</td>\n",
       "      <td>GT:GQ:DP:RO:QR:AO:QA:GL</td>\n",
       "      <td>0/1:150.423:53:28:1054:25:905:-80.8135,0,-94.238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114354</th>\n",
       "      <td>000000F</td>\n",
       "      <td>33303436</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>378.56</td>\n",
       "      <td>.</td>\n",
       "      <td>AB=0.340909;ABP=12.6832;AC=1;AF=0.5;AN=2;AO=15...</td>\n",
       "      <td>GT:GQ:DP:RO:QR:AO:QA:GL</td>\n",
       "      <td>0/1:160.002:44:29:1029:15:514:-44.7189,0,-91.0811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114355</th>\n",
       "      <td>000000F</td>\n",
       "      <td>33303793</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>393.46</td>\n",
       "      <td>.</td>\n",
       "      <td>AB=0.384615;ABP=7.52028;AC=1;AF=0.5;AN=2;AO=15...</td>\n",
       "      <td>GT:GQ:DP:RO:QR:AO:QA:GL</td>\n",
       "      <td>0/1:160.002:39:24:843:15:511:-44.9909,0,-74.8815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114356</th>\n",
       "      <td>000000F</td>\n",
       "      <td>33303946</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>512.48</td>\n",
       "      <td>.</td>\n",
       "      <td>AB=0.459459;ABP=3.5385;AC=1;AF=0.5;AN=2;AO=17;...</td>\n",
       "      <td>GT:GQ:DP:RO:QR:AO:QA:GL</td>\n",
       "      <td>0/1:150.795:37:20:720:17:634:-56.4964,0,-64.2234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114357</th>\n",
       "      <td>000000F</td>\n",
       "      <td>33304392</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>878.91</td>\n",
       "      <td>.</td>\n",
       "      <td>AB=0.697674;ABP=17.6046;AC=1;AF=0.5;AN=2;AO=30...</td>\n",
       "      <td>GT:GQ:DP:RO:QR:AO:QA:GL</td>\n",
       "      <td>0/1:147.573:43:13:473:30:1072:-94.4562,0,-40.5528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114358</th>\n",
       "      <td>000000F</td>\n",
       "      <td>33304496</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>522.30</td>\n",
       "      <td>.</td>\n",
       "      <td>AB=0.703704;ABP=12.7417;AC=1;AF=0.5;AN=2;AO=19...</td>\n",
       "      <td>GT:GQ:DP:RO:QR:AO:QA:GL</td>\n",
       "      <td>0/1:160.002:27:8:304:19:662:-58.147,0,-25.9586</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114359</th>\n",
       "      <td>000000F</td>\n",
       "      <td>33304598</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>416.39</td>\n",
       "      <td>.</td>\n",
       "      <td>AB=0.482759;ABP=3.08518;AC=1;AF=0.5;AN=2;AO=14...</td>\n",
       "      <td>GT:GQ:DP:RO:QR:AO:QA:GL</td>\n",
       "      <td>0/1:151.018:29:15:528:14:520:-46.3312,0,-47.0318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114360</th>\n",
       "      <td>000000F</td>\n",
       "      <td>33304869</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>790.75</td>\n",
       "      <td>.</td>\n",
       "      <td>AB=0.619048;ABP=8.18047;AC=1;AF=0.5;AN=2;AO=26...</td>\n",
       "      <td>GT:GQ:DP:RO:QR:AO:QA:GL</td>\n",
       "      <td>0/1:153.842:42:16:589:26:954:-84.8051,0,-51.9563</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114361</th>\n",
       "      <td>000000F</td>\n",
       "      <td>33304986</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>241.04</td>\n",
       "      <td>.</td>\n",
       "      <td>AB=0.307692;ABP=15.538;AC=1;AF=0.5;AN=2;AO=12;...</td>\n",
       "      <td>GT:GQ:DP:RO:QR:AO:QA:GL</td>\n",
       "      <td>0/1:144.554:39:27:862:12:379:-32.2779,0,-75.7514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114362</th>\n",
       "      <td>000000F</td>\n",
       "      <td>33305056</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>249.36</td>\n",
       "      <td>.</td>\n",
       "      <td>AB=0.27907;ABP=21.2406;AC=1;AF=0.5;AN=2;AO=12;...</td>\n",
       "      <td>GT:GQ:DP:RO:QR:AO:QA:GL</td>\n",
       "      <td>0/1:160.002:43:31:1045:12:412:-34.6648,0,-91.6286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114363</th>\n",
       "      <td>000000F</td>\n",
       "      <td>33305066</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>160.26</td>\n",
       "      <td>.</td>\n",
       "      <td>AB=0.25;ABP=20.3821;AC=1;AF=0.5;AN=2;AO=8;CIGA...</td>\n",
       "      <td>GT:GQ:DP:RO:QR:AO:QA:GL</td>\n",
       "      <td>0/1:143.184:32:24:828:8:297:-24.4902,0,-72.254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114364</th>\n",
       "      <td>000000F</td>\n",
       "      <td>33305071</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>81.31</td>\n",
       "      <td>.</td>\n",
       "      <td>AB=0.222222;ABP=21.1059;AC=1;AF=0.5;AN=2;AO=6;...</td>\n",
       "      <td>GT:GQ:DP:RO:QR:AO:QA:GL</td>\n",
       "      <td>0/1:81.3097:27:21:742:6:194:-15.1268,0,-64.4768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114365</th>\n",
       "      <td>000000F</td>\n",
       "      <td>33305769</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>772.34</td>\n",
       "      <td>.</td>\n",
       "      <td>AB=0.5625;ABP=4.6389;AC=1;AF=0.5;AN=2;AO=27;CI...</td>\n",
       "      <td>GT:GQ:DP:RO:QR:AO:QA:GL</td>\n",
       "      <td>0/1:148.765:48:21:734:27:938:-83.6665,0,-65.3087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114366</th>\n",
       "      <td>000000F</td>\n",
       "      <td>33305791</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>750.60</td>\n",
       "      <td>.</td>\n",
       "      <td>AB=0.543478;ABP=3.76559;AC=1;AF=0.5;AN=2;AO=25...</td>\n",
       "      <td>GT:GQ:DP:RO:QR:AO:QA:GL</td>\n",
       "      <td>0/1:160.002:46:21:752:25:906:-80.8966,0,-67.0323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114367</th>\n",
       "      <td>000000F</td>\n",
       "      <td>33306087</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>406.29</td>\n",
       "      <td>.</td>\n",
       "      <td>AB=0.46875;ABP=3.28173;AC=1;AF=0.5;AN=2;AO=15;...</td>\n",
       "      <td>GT:GQ:DP:RO:QR:AO:QA:GL</td>\n",
       "      <td>0/1:149.028:32:17:605:15:520:-46.2663,0,-53.9255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114368</th>\n",
       "      <td>000000F</td>\n",
       "      <td>33307001</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>202.08</td>\n",
       "      <td>.</td>\n",
       "      <td>AB=0.272727;ABP=17.8158;AC=1;AF=0.5;AN=2;AO=9;...</td>\n",
       "      <td>GT:GQ:DP:RO:QR:AO:QA:GL</td>\n",
       "      <td>0/1:160.002:33:24:824:9:318:-26.6256,0,-72.1556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114369</th>\n",
       "      <td>000000F</td>\n",
       "      <td>33307014</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>224.45</td>\n",
       "      <td>.</td>\n",
       "      <td>AB=0.310345;ABP=12.0706;AC=1;AF=0.5;AN=2;AO=9;...</td>\n",
       "      <td>GT:GQ:DP:RO:QR:AO:QA:GL</td>\n",
       "      <td>0/1:151.018:29:20:695:9:323:-27.6997,0,-61.1683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114370</th>\n",
       "      <td>000000F</td>\n",
       "      <td>33307318</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>391.34</td>\n",
       "      <td>.</td>\n",
       "      <td>AB=0.388889;ABP=6.8707;AC=1;AF=0.5;AN=2;AO=14;...</td>\n",
       "      <td>GT:GQ:DP:RO:QR:AO:QA:GL</td>\n",
       "      <td>0/1:160.002:36:22:803:14:499:-44.0087,0,-71.3773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114371</th>\n",
       "      <td>000000F</td>\n",
       "      <td>33307341</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>378.72</td>\n",
       "      <td>.</td>\n",
       "      <td>AB=0.368421;ABP=8.7247;AC=1;AF=0.5;AN=2;AO=14;...</td>\n",
       "      <td>GT:GQ:DP:RO:QR:AO:QA:GL</td>\n",
       "      <td>0/1:160.002:38:24:866:14:492:-43.1777,0,-76.8471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114372</th>\n",
       "      <td>000000F</td>\n",
       "      <td>33307556</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>778.89</td>\n",
       "      <td>.</td>\n",
       "      <td>AB=0.642857;ABP=10.4553;AC=1;AF=0.5;AN=2;AO=27...</td>\n",
       "      <td>GT:GQ:DP:RO:QR:AO:QA:GL</td>\n",
       "      <td>0/1:153.842:42:15:548:27:951:-84.2932,0,-48.0363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114373</th>\n",
       "      <td>000000F</td>\n",
       "      <td>33307616</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>653.44</td>\n",
       "      <td>.</td>\n",
       "      <td>AB=0.605263;ABP=6.66752;AC=1;AF=0.5;AN=2;AO=23...</td>\n",
       "      <td>GT:GQ:DP:RO:QR:AO:QA:GL</td>\n",
       "      <td>0/1:160.002:38:15:548:23:799:-71.0078,0,-48.4357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114374</th>\n",
       "      <td>000000F</td>\n",
       "      <td>33308071</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>446.50</td>\n",
       "      <td>.</td>\n",
       "      <td>AB=0.444444;ABP=3.9754;AC=1;AF=0.5;AN=2;AO=16;...</td>\n",
       "      <td>GT:GQ:DP:RO:QR:AO:QA:GL</td>\n",
       "      <td>0/1:160.002:36:20:706:16:567:-50.4111,0,-62.9197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114375</th>\n",
       "      <td>000000F</td>\n",
       "      <td>33308537</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>559.46</td>\n",
       "      <td>.</td>\n",
       "      <td>AB=0.542857;ABP=3.56868;AC=1;AF=0.5;AN=2;AO=19...</td>\n",
       "      <td>GT:GQ:DP:RO:QR:AO:QA:GL</td>\n",
       "      <td>0/1:147.673:35:16:575:19:692:-61.7167,0,-51.1818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114376</th>\n",
       "      <td>000000F</td>\n",
       "      <td>33308554</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>623.20</td>\n",
       "      <td>.</td>\n",
       "      <td>AB=0.567568;ABP=4.47751;AC=1;AF=0.5;AN=2;AO=21...</td>\n",
       "      <td>GT:GQ:DP:RO:QR:AO:QA:GL</td>\n",
       "      <td>0/1:150.804:37:16:586:21:762:-67.9145,0,-52.0779</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114377</th>\n",
       "      <td>000000F</td>\n",
       "      <td>33308940</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>904.04</td>\n",
       "      <td>.</td>\n",
       "      <td>AB=0.588235;ABP=6.45911;AC=1;AF=0.5;AN=2;AO=30...</td>\n",
       "      <td>GT:GQ:DP:RO:QR:AO:QA:GL</td>\n",
       "      <td>0/1:160.002:51:21:747:30:1077:-95.9951,0,-66.2918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114378</th>\n",
       "      <td>000000F</td>\n",
       "      <td>33309047</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>568.73</td>\n",
       "      <td>.</td>\n",
       "      <td>AB=0.463415;ABP=3.48696;AC=1;AF=0.5;AN=2;AO=19...</td>\n",
       "      <td>GT:GQ:DP:RO:QR:AO:QA:GL</td>\n",
       "      <td>0/1:160.002:41:22:792:19:695:-61.9621,0,-70.6863</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114379</th>\n",
       "      <td>000000F</td>\n",
       "      <td>33309092</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>502.53</td>\n",
       "      <td>.</td>\n",
       "      <td>AB=0.515152;ABP=3.0761;AC=1;AF=0.5;AN=2;AO=17;...</td>\n",
       "      <td>GT:GQ:DP:RO:QR:AO:QA:GL</td>\n",
       "      <td>0/1:160.002:33:16:581:17:616:-54.9354,0,-51.7861</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114380</th>\n",
       "      <td>000000F</td>\n",
       "      <td>33309159</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>343.83</td>\n",
       "      <td>.</td>\n",
       "      <td>AB=0.541667;ABP=3.37221;AC=1;AF=0.5;AN=2;AO=13...</td>\n",
       "      <td>GT:GQ:DP:RO:QR:AO:QA:GL</td>\n",
       "      <td>0/1:160.002:24:11:359:13:439:-39.0202,0,-31.8089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114381</th>\n",
       "      <td>000000F</td>\n",
       "      <td>33309197</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>375.37</td>\n",
       "      <td>.</td>\n",
       "      <td>AB=0.56;ABP=3.79203;AC=1;AF=0.5;AN=2;AO=14;CIG...</td>\n",
       "      <td>GT:GQ:DP:RO:QR:AO:QA:GL</td>\n",
       "      <td>0/1:160.002:25:11:367:14:473:-42.0312,0,-32.487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114382</th>\n",
       "      <td>000000F</td>\n",
       "      <td>33309844</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>382.66</td>\n",
       "      <td>.</td>\n",
       "      <td>AB=0.454545;ABP=3.60252;AC=1;AF=0.5;AN=2;AO=15...</td>\n",
       "      <td>GT:GQ:DP:RO:QR:AO:QA:GL</td>\n",
       "      <td>0/1:160.002:33:18:606:15:491:-43.5992,0,-53.9585</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>114383 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              0         1          2     3     4        5  6                                                  7                        8                                                  9\n",
       "0       000000F        43   methyl-1     X    MX   100.00  .                            MEAN1=2.373;MEAN2=3.243                       GT                                                0/1\n",
       "1       000000F       125   methyl-2     X    MX   100.00  .                            MEAN1=3.087;MEAN2=2.231                       GT                                                0/1\n",
       "2       000000F       613   methyl-3     X    MX   100.00  .                            MEAN1=3.655;MEAN2=2.698                       GT                                                0/1\n",
       "3       000000F       647   methyl-4     X    MX   100.00  .                            MEAN1=2.993;MEAN2=2.119                       GT                                                0/1\n",
       "4       000000F       966   methyl-5     X    MX   100.00  .                            MEAN1=3.628;MEAN2=2.617                       GT                                                0/1\n",
       "5       000000F      1281   methyl-6     X    MX   100.00  .                            MEAN1=3.102;MEAN2=2.553                       GT                                                0/1\n",
       "6       000000F      1520   methyl-7     X    MX   100.00  .                            MEAN1=2.954;MEAN2=2.018                       GT                                                0/1\n",
       "7       000000F      1581   methyl-8     X    MX   100.00  .                            MEAN1=3.041;MEAN2=1.927                       GT                                                0/1\n",
       "8       000000F      1582   methyl-9     X    MX   100.00  .                            MEAN1=3.193;MEAN2=2.476                       GT                                                0/1\n",
       "9       000000F      1735  methyl-10     X    MX   100.00  .                            MEAN1=2.553;MEAN2=3.254                       GT                                                0/1\n",
       "10      000000F      1804  methyl-11     X    MX   100.00  .                            MEAN1=3.662;MEAN2=4.263                       GT                                                0/1\n",
       "11      000000F      1870  methyl-12     X    MX   100.00  .                            MEAN1=3.015;MEAN2=2.489                       GT                                                0/1\n",
       "12      000000F      2001          .  ATTA  CATT  1714.17  .  AB=0;ABP=0;AC=2;AF=1;AN=2;AO=53;CIGAR=2X1M1X;D...  GT:GQ:DP:RO:QR:AO:QA:GL     1/1:160.002:53:0:0:53:1948:-175.688,-15.9546,0\n",
       "13      000000F      2328  methyl-13     X    MX   100.00  .                             MEAN1=3.335;MEAN2=2.75                       GT                                                0/1\n",
       "14      000000F      2363  methyl-14     X    MX   100.00  .                            MEAN1=2.501;MEAN2=2.997                       GT                                                0/1\n",
       "15      000000F      2387  methyl-15     X    MX   100.00  .                            MEAN1=3.328;MEAN2=2.507                       GT                                                0/1\n",
       "16      000000F      2675  methyl-16     X    MX   100.00  .                            MEAN1=2.513;MEAN2=2.872                       GT                                                0/1\n",
       "17      000000F      3089  methyl-17     X    MX   100.00  .                             MEAN1=3.296;MEAN2=2.48                       GT                                                0/1\n",
       "18      000000F      3638  methyl-18     X    MX   100.00  .                            MEAN1=3.161;MEAN2=2.709                       GT                                                0/1\n",
       "19      000000F      3741  methyl-19     X    MX   100.00  .                            MEAN1=2.692;MEAN2=3.728                       GT                                                0/1\n",
       "20      000000F      4284  methyl-20     X    MX   100.00  .                            MEAN1=3.013;MEAN2=2.549                       GT                                                0/1\n",
       "21      000000F      4371  methyl-21     X    MX   100.00  .                            MEAN1=2.792;MEAN2=2.477                       GT                                                0/1\n",
       "22      000000F      4645  methyl-22     X    MX   100.00  .                            MEAN1=2.267;MEAN2=2.932                       GT                                                0/1\n",
       "23      000000F      4648  methyl-23     X    MX   100.00  .                            MEAN1=2.876;MEAN2=3.478                       GT                                                0/1\n",
       "24      000000F      5164  methyl-24     X    MX   100.00  .                            MEAN1=2.692;MEAN2=3.056                       GT                                                0/1\n",
       "25      000000F      5351  methyl-25     X    MX   100.00  .                            MEAN1=3.278;MEAN2=2.582                       GT                                                0/1\n",
       "26      000000F      5369          .     A     G   805.85  .  AB=0.530612;ABP=3.40914;AC=1;AF=0.5;AN=2;AO=26...  GT:GQ:DP:RO:QR:AO:QA:GL   0/1:160.002:49:23:839:26:969:-86.5982,0,-74.8903\n",
       "27      000000F      6153  methyl-26     X    MX   100.00  .                            MEAN1=3.199;MEAN2=2.701                       GT                                                0/1\n",
       "28      000000F      6259  methyl-27     X    MX   100.00  .                            MEAN1=3.066;MEAN2=2.208                       GT                                                0/1\n",
       "29      000000F      6310  methyl-28     X    MX   100.00  .                             MEAN1=2.467;MEAN2=1.85                       GT                                                0/1\n",
       "...         ...       ...        ...   ...   ...      ... ..                                                ...                      ...                                                ...\n",
       "114353  000000F  33303233          .     T     G   754.60  .  AB=0.471698;ABP=3.37904;AC=1;AF=0.5;AN=2;AO=25...  GT:GQ:DP:RO:QR:AO:QA:GL   0/1:150.423:53:28:1054:25:905:-80.8135,0,-94.238\n",
       "114354  000000F  33303436          .     T     G   378.56  .  AB=0.340909;ABP=12.6832;AC=1;AF=0.5;AN=2;AO=15...  GT:GQ:DP:RO:QR:AO:QA:GL  0/1:160.002:44:29:1029:15:514:-44.7189,0,-91.0811\n",
       "114355  000000F  33303793          .     T     C   393.46  .  AB=0.384615;ABP=7.52028;AC=1;AF=0.5;AN=2;AO=15...  GT:GQ:DP:RO:QR:AO:QA:GL   0/1:160.002:39:24:843:15:511:-44.9909,0,-74.8815\n",
       "114356  000000F  33303946          .     C     T   512.48  .  AB=0.459459;ABP=3.5385;AC=1;AF=0.5;AN=2;AO=17;...  GT:GQ:DP:RO:QR:AO:QA:GL   0/1:150.795:37:20:720:17:634:-56.4964,0,-64.2234\n",
       "114357  000000F  33304392          .     G     A   878.91  .  AB=0.697674;ABP=17.6046;AC=1;AF=0.5;AN=2;AO=30...  GT:GQ:DP:RO:QR:AO:QA:GL  0/1:147.573:43:13:473:30:1072:-94.4562,0,-40.5528\n",
       "114358  000000F  33304496          .     G     A   522.30  .  AB=0.703704;ABP=12.7417;AC=1;AF=0.5;AN=2;AO=19...  GT:GQ:DP:RO:QR:AO:QA:GL     0/1:160.002:27:8:304:19:662:-58.147,0,-25.9586\n",
       "114359  000000F  33304598          .     A     G   416.39  .  AB=0.482759;ABP=3.08518;AC=1;AF=0.5;AN=2;AO=14...  GT:GQ:DP:RO:QR:AO:QA:GL   0/1:151.018:29:15:528:14:520:-46.3312,0,-47.0318\n",
       "114360  000000F  33304869          .     T     C   790.75  .  AB=0.619048;ABP=8.18047;AC=1;AF=0.5;AN=2;AO=26...  GT:GQ:DP:RO:QR:AO:QA:GL   0/1:153.842:42:16:589:26:954:-84.8051,0,-51.9563\n",
       "114361  000000F  33304986          .     C     T   241.04  .  AB=0.307692;ABP=15.538;AC=1;AF=0.5;AN=2;AO=12;...  GT:GQ:DP:RO:QR:AO:QA:GL   0/1:144.554:39:27:862:12:379:-32.2779,0,-75.7514\n",
       "114362  000000F  33305056          .     G     A   249.36  .  AB=0.27907;ABP=21.2406;AC=1;AF=0.5;AN=2;AO=12;...  GT:GQ:DP:RO:QR:AO:QA:GL  0/1:160.002:43:31:1045:12:412:-34.6648,0,-91.6286\n",
       "114363  000000F  33305066          .     T     C   160.26  .  AB=0.25;ABP=20.3821;AC=1;AF=0.5;AN=2;AO=8;CIGA...  GT:GQ:DP:RO:QR:AO:QA:GL     0/1:143.184:32:24:828:8:297:-24.4902,0,-72.254\n",
       "114364  000000F  33305071          .     A     G    81.31  .  AB=0.222222;ABP=21.1059;AC=1;AF=0.5;AN=2;AO=6;...  GT:GQ:DP:RO:QR:AO:QA:GL    0/1:81.3097:27:21:742:6:194:-15.1268,0,-64.4768\n",
       "114365  000000F  33305769          .     A     G   772.34  .  AB=0.5625;ABP=4.6389;AC=1;AF=0.5;AN=2;AO=27;CI...  GT:GQ:DP:RO:QR:AO:QA:GL   0/1:148.765:48:21:734:27:938:-83.6665,0,-65.3087\n",
       "114366  000000F  33305791          .     C     T   750.60  .  AB=0.543478;ABP=3.76559;AC=1;AF=0.5;AN=2;AO=25...  GT:GQ:DP:RO:QR:AO:QA:GL   0/1:160.002:46:21:752:25:906:-80.8966,0,-67.0323\n",
       "114367  000000F  33306087          .     C     T   406.29  .  AB=0.46875;ABP=3.28173;AC=1;AF=0.5;AN=2;AO=15;...  GT:GQ:DP:RO:QR:AO:QA:GL   0/1:149.028:32:17:605:15:520:-46.2663,0,-53.9255\n",
       "114368  000000F  33307001          .     C     T   202.08  .  AB=0.272727;ABP=17.8158;AC=1;AF=0.5;AN=2;AO=9;...  GT:GQ:DP:RO:QR:AO:QA:GL    0/1:160.002:33:24:824:9:318:-26.6256,0,-72.1556\n",
       "114369  000000F  33307014          .     G     T   224.45  .  AB=0.310345;ABP=12.0706;AC=1;AF=0.5;AN=2;AO=9;...  GT:GQ:DP:RO:QR:AO:QA:GL    0/1:151.018:29:20:695:9:323:-27.6997,0,-61.1683\n",
       "114370  000000F  33307318          .     C     T   391.34  .  AB=0.388889;ABP=6.8707;AC=1;AF=0.5;AN=2;AO=14;...  GT:GQ:DP:RO:QR:AO:QA:GL   0/1:160.002:36:22:803:14:499:-44.0087,0,-71.3773\n",
       "114371  000000F  33307341          .     A     C   378.72  .  AB=0.368421;ABP=8.7247;AC=1;AF=0.5;AN=2;AO=14;...  GT:GQ:DP:RO:QR:AO:QA:GL   0/1:160.002:38:24:866:14:492:-43.1777,0,-76.8471\n",
       "114372  000000F  33307556          .     C     T   778.89  .  AB=0.642857;ABP=10.4553;AC=1;AF=0.5;AN=2;AO=27...  GT:GQ:DP:RO:QR:AO:QA:GL   0/1:153.842:42:15:548:27:951:-84.2932,0,-48.0363\n",
       "114373  000000F  33307616          .     T     C   653.44  .  AB=0.605263;ABP=6.66752;AC=1;AF=0.5;AN=2;AO=23...  GT:GQ:DP:RO:QR:AO:QA:GL   0/1:160.002:38:15:548:23:799:-71.0078,0,-48.4357\n",
       "114374  000000F  33308071          .     G     A   446.50  .  AB=0.444444;ABP=3.9754;AC=1;AF=0.5;AN=2;AO=16;...  GT:GQ:DP:RO:QR:AO:QA:GL   0/1:160.002:36:20:706:16:567:-50.4111,0,-62.9197\n",
       "114375  000000F  33308537          .     G     C   559.46  .  AB=0.542857;ABP=3.56868;AC=1;AF=0.5;AN=2;AO=19...  GT:GQ:DP:RO:QR:AO:QA:GL   0/1:147.673:35:16:575:19:692:-61.7167,0,-51.1818\n",
       "114376  000000F  33308554          .     T     C   623.20  .  AB=0.567568;ABP=4.47751;AC=1;AF=0.5;AN=2;AO=21...  GT:GQ:DP:RO:QR:AO:QA:GL   0/1:150.804:37:16:586:21:762:-67.9145,0,-52.0779\n",
       "114377  000000F  33308940          .     A     G   904.04  .  AB=0.588235;ABP=6.45911;AC=1;AF=0.5;AN=2;AO=30...  GT:GQ:DP:RO:QR:AO:QA:GL  0/1:160.002:51:21:747:30:1077:-95.9951,0,-66.2918\n",
       "114378  000000F  33309047          .     A     G   568.73  .  AB=0.463415;ABP=3.48696;AC=1;AF=0.5;AN=2;AO=19...  GT:GQ:DP:RO:QR:AO:QA:GL   0/1:160.002:41:22:792:19:695:-61.9621,0,-70.6863\n",
       "114379  000000F  33309092          .     A     G   502.53  .  AB=0.515152;ABP=3.0761;AC=1;AF=0.5;AN=2;AO=17;...  GT:GQ:DP:RO:QR:AO:QA:GL   0/1:160.002:33:16:581:17:616:-54.9354,0,-51.7861\n",
       "114380  000000F  33309159          .     A     G   343.83  .  AB=0.541667;ABP=3.37221;AC=1;AF=0.5;AN=2;AO=13...  GT:GQ:DP:RO:QR:AO:QA:GL   0/1:160.002:24:11:359:13:439:-39.0202,0,-31.8089\n",
       "114381  000000F  33309197          .     T     C   375.37  .  AB=0.56;ABP=3.79203;AC=1;AF=0.5;AN=2;AO=14;CIG...  GT:GQ:DP:RO:QR:AO:QA:GL    0/1:160.002:25:11:367:14:473:-42.0312,0,-32.487\n",
       "114382  000000F  33309844          .     T     C   382.66  .  AB=0.454545;ABP=3.60252;AC=1;AF=0.5;AN=2;AO=15...  GT:GQ:DP:RO:QR:AO:QA:GL   0/1:160.002:33:18:606:15:491:-43.5992,0,-53.9585\n",
       "\n",
       "[114383 rows x 10 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mergedvcf = mergedvcf.reset_index(drop=True)\n",
    "mergedvcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for bid, block in block_reader.blocks.iteritems():\n",
    "    block_variants = [(v.pos, v.r_allele, v.v_allele, int(v.hap1), int(v.hap2)) for v \\\n",
    "              in block.variants.values()]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
