{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from math import log\n",
    "unit_list = zip(['bytes', 'kB', 'MB', 'GB', 'TB', 'PB'], [0, 0, 1, 2, 2, 2])\n",
    "def sizeof_fmt(num):\n",
    "    \"\"\"Human friendly file size\"\"\"\n",
    "    if num > 1:\n",
    "        exponent = min(int(log(num, 1024)), len(unit_list) - 1)\n",
    "        quotient = float(num) / 1024**exponent\n",
    "        unit, num_decimals = unit_list[exponent]\n",
    "        format_string = '{:.%sf} {}' % (num_decimals)\n",
    "        return format_string.format(quotient, unit)\n",
    "    if num == 0:\n",
    "        return '0 bytes'\n",
    "    if num == 1:\n",
    "        return '1 byte'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "class ListTable(list):\n",
    "    \"\"\" Overridden list class which takes a 2-dimensional list of \n",
    "        the form [[1,2,3],[4,5,6]], and renders an HTML Table in \n",
    "        IPython Notebook. \"\"\"\n",
    "    \n",
    "    def _repr_html_(self):\n",
    "        html = [\"<table>\"]\n",
    "        count = 0\n",
    "        for row in self:\n",
    "            html.append(\"<tr>\")\n",
    "            \n",
    "            for col in row:\n",
    "                if count == 0:\n",
    "                    html.append(\"<td><strong>{0}</strong></td>\".format(col))\n",
    "                else:\n",
    "                    html.append(\"<td>{0}</td>\".format(col))\n",
    "            \n",
    "            html.append(\"</tr>\")\n",
    "            count += 1\n",
    "        html.append(\"</table>\")\n",
    "        return ''.join(html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Imports / style (run this first always)\n",
    "\n",
    "%matplotlib inline\n",
    "from IPython.display import FileLink, FileLinks\n",
    "from IPython.core import display\n",
    "from collections import defaultdict\n",
    "import json\n",
    "import sys\n",
    "import time\n",
    "\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "\n",
    "from matplotlib import rcParams\n",
    "import matplotlib.cm as cm\n",
    "import matplotlib as mpl\n",
    "\n",
    "class AwesomeError(Exception):\n",
    "     def __init__(self, value):\n",
    "         self.value = value\n",
    "         pass\n",
    "     def __str__(self):\n",
    "         return repr(self.value)\n",
    "         pass\n",
    "\n",
    "#colorbrewer2 Dark2 qualitative color table\n",
    "dark2_colors = [(0.10588235294117647, 0.6196078431372549, 0.4666666666666667),\n",
    "                (0.8509803921568627, 0.37254901960784315, 0.00784313725490196),\n",
    "                (0.4588235294117647, 0.4392156862745098, 0.7019607843137254),\n",
    "                (0.9058823529411765, 0.1607843137254902, 0.5411764705882353),\n",
    "                (0.4, 0.6509803921568628, 0.11764705882352941),\n",
    "                (0.9019607843137255, 0.6705882352941176, 0.00784313725490196),\n",
    "                (0.6509803921568628, 0.4627450980392157, 0.11372549019607843)]\n",
    "\n",
    "rcParams['figure.figsize'] = (10, 6)\n",
    "rcParams['figure.dpi'] = 150\n",
    "rcParams['axes.color_cycle'] = dark2_colors\n",
    "rcParams['lines.linewidth'] = 2\n",
    "rcParams['axes.facecolor'] = 'white'\n",
    "rcParams['font.size'] = 14\n",
    "rcParams['patch.edgecolor'] = 'white'\n",
    "rcParams['patch.facecolor'] = dark2_colors[0]\n",
    "rcParams['font.family'] = 'StixGeneral'\n",
    "\n",
    "\n",
    "def remove_border(axes=None, top=False, right=False, left=True, bottom=True):\n",
    "    \"\"\"\n",
    "    Minimize chartjunk by stripping out unnecesasry plot borders and axis ticks\n",
    "    \n",
    "    The top/right/left/bottom keywords toggle whether the corresponding plot border is drawn\n",
    "    \"\"\"\n",
    "    ax = axes or plt.gca()\n",
    "    ax.spines['top'].set_visible(top)\n",
    "    ax.spines['right'].set_visible(right)\n",
    "    ax.spines['left'].set_visible(left)\n",
    "    ax.spines['bottom'].set_visible(bottom)\n",
    "    \n",
    "    #turn off all ticks\n",
    "    ax.yaxis.set_ticks_position('none')\n",
    "    ax.xaxis.set_ticks_position('none')\n",
    "    \n",
    "    #now re-enable visibles\n",
    "    if top:\n",
    "        ax.xaxis.tick_top()\n",
    "    if bottom:\n",
    "        ax.xaxis.tick_bottom()\n",
    "    if left:\n",
    "        ax.yaxis.tick_left()\n",
    "    if right:\n",
    "        ax.yaxis.tick_right()\n",
    "        \n",
    "pd.set_option('display.width', 500)\n",
    "pd.set_option('display.max_columns', 100)\n",
    "\n",
    "import Bio as bp\n",
    "from Bio.Sequencing.Applications import BwaAlignCommandline as bwa_aln\n",
    "from Bio.Sequencing.Applications import BwaSamseCommandline as bwa_samse\n",
    "from Bio.Sequencing.Applications import BwaSampeCommandline as bwa_sampe\n",
    "from Bio.Sequencing.Applications import BwaIndexCommandline as bwa_index\n",
    "from Bio.Sequencing.Applications import BwaBwaswCommandline as bwa_bwasw\n",
    "import HTSeq as ht\n",
    "import subprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[33mYou are using pip version 6.0.8, however version 7.1.0 is available.\n",
      "You should consider upgrading via the 'pip install --upgrade pip' command.\u001b[0m\n",
      "Collecting subprocess\n",
      "\u001b[31m  Could not find any downloads that satisfy the requirement subprocess\u001b[0m\n",
      "\u001b[31m  No distributions at all found for subprocess\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!pip install subprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "root prints out directories only from what you specified\n",
      "dirs prints out sub-directories from root\n",
      "files prints out all files from root and directories\n",
      "********************\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import struct\n",
    "\n",
    "def getuncompressedsize(filename):\n",
    "    with open(filename) as f:\n",
    "        f.seek(-4, 2)\n",
    "        return struct.unpack('I', f.read(4))[0]\n",
    "    \n",
    "print \"root prints out directories only from what you specified\"\n",
    "print \"dirs prints out sub-directories from root\"\n",
    "print \"files prints out all files from root and directories\"\n",
    "print \"*\" * 20\n",
    "master_file_array = []\n",
    "prefix = \"/home/neffr01/minerva/short_reads/\"\n",
    "for root, dirs, files in os.walk(prefix):\n",
    "    #print root\n",
    "    #print dirs\n",
    "    holdit = []\n",
    "    for file_ in files:\n",
    "        tmp = os.path.join(root, file_)\n",
    "        try:\n",
    "            size = os.path.getsize(tmp)\n",
    "        except:\n",
    "            continue\n",
    "        ext = tmp.split('/')[-1].split('.')[-1]\n",
    "        if ext == \"gz\":\n",
    "            ext = '.'.join([tmp.split('/')[-1].split('.')[-2], 'gz'])\n",
    "            #size = getuncompressedsize(tmp)\n",
    "        holdit.append((tmp.split('/')[5], file_, ext, size, tmp))\n",
    "    master_file_array.append(holdit)\n",
    "    #print files\n",
    "len(master_file_array)\n",
    "master_file_array2 = [item for sublist in master_file_array for item in sublist] # only do this once!\n",
    "master_file_array2.sort()\n",
    "pdarr = pd.DataFrame(master_file_array2)\n",
    "pdarr_pretty = pd.DataFrame(master_file_array2, columns=[\"folder\", \"name\", \"extension\", \"size\", \"path\"])[['.' not in i[0] for i in pdarr[1]]]\n",
    "pdarr = pdarr[['.' not in i[0] for i in pdarr[1]]] # remove hidden files\n",
    "grouped = pdarr.groupby(pdarr[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import struct\n",
    "def getlastchars(filename):\n",
    "    with open(filename) as f:\n",
    "        f.seek(-16, 2)\n",
    "        return f.read(16)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "fastq_files = pdarr_pretty[(pdarr_pretty['extension'] == 'fastq.gz')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/neffr01/anaconda/lib/python2.7/site-packages/IPython/kernel/__main__.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  if __name__ == '__main__':\n"
     ]
    }
   ],
   "source": [
    "fastq_files['parent'] = fastq_files['path'].map(lambda x: '/'.join(x.split('/')[-4:-3]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>folder</th>\n",
       "      <th>name</th>\n",
       "      <th>extension</th>\n",
       "      <th>size</th>\n",
       "      <th>path</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>82 </th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2A1_CGATGT_L001_R1_001.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 428893454</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140528_D00360_0019_BH8VDAADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83 </th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2A1_CGATGT_L001_R1_001.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 509951892</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140528_D00360_0018_AH8VC6ADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84 </th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2A1_CGATGT_L001_R1_001.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 513979404</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140605_D00360_0021_BH9V1VADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85 </th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2A1_CGATGT_L001_R1_001.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 514453620</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140605_D00360_0020_AH9V1RADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86 </th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2A1_CGATGT_L001_R1_001.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 523477905</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140609_D00360_0022_AH9UJNADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87 </th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2A1_CGATGT_L001_R1_001.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 524462927</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140609_D00360_0023_BH9UD5ADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88 </th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2A1_CGATGT_L001_R1_002.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 167071275</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140605_D00360_0020_AH9V1RADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89 </th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2A1_CGATGT_L001_R1_002.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 180252467</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140605_D00360_0021_BH9V1VADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90 </th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2A1_CGATGT_L001_R1_002.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 203174069</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140528_D00360_0018_AH8VC6ADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91 </th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2A1_CGATGT_L001_R1_002.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 206957195</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140609_D00360_0023_BH9UD5ADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92 </th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2A1_CGATGT_L001_R1_002.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 225431818</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140609_D00360_0022_AH9UJNADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93 </th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2A1_CGATGT_L001_R2_001.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 459671519</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140528_D00360_0019_BH8VDAADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94 </th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2A1_CGATGT_L001_R2_001.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 548917388</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140605_D00360_0021_BH9V1VADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95 </th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2A1_CGATGT_L001_R2_001.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 549026753</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140528_D00360_0018_AH8VC6ADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96 </th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2A1_CGATGT_L001_R2_001.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 554220163</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140605_D00360_0020_AH9V1RADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97 </th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2A1_CGATGT_L001_R2_001.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 556144016</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140609_D00360_0023_BH9UD5ADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98 </th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2A1_CGATGT_L001_R2_001.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 558736324</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140609_D00360_0022_AH9UJNADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99 </th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2A1_CGATGT_L001_R2_002.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 178949495</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140605_D00360_0020_AH9V1RADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2A1_CGATGT_L001_R2_002.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 192699874</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140605_D00360_0021_BH9V1VADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2A1_CGATGT_L001_R2_002.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 218221790</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140528_D00360_0018_AH8VC6ADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2A1_CGATGT_L001_R2_002.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 219959720</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140609_D00360_0023_BH9UD5ADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2A1_CGATGT_L001_R2_002.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 239419065</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140609_D00360_0022_AH9UJNADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2A1_CGATGT_L002_R1_001.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 436819051</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140528_D00360_0019_BH8VDAADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2A1_CGATGT_L002_R1_001.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 509833118</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140528_D00360_0018_AH8VC6ADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2A1_CGATGT_L002_R1_001.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 514672669</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140605_D00360_0020_AH9V1RADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2A1_CGATGT_L002_R1_001.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 516033114</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140605_D00360_0021_BH9V1VADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2A1_CGATGT_L002_R1_001.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 523301733</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140609_D00360_0022_AH9UJNADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2A1_CGATGT_L002_R1_001.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 526086768</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140609_D00360_0023_BH9UD5ADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2A1_CGATGT_L002_R1_002.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 172475775</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140605_D00360_0021_BH9V1VADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2A1_CGATGT_L002_R1_002.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 179401358</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140605_D00360_0020_AH9V1RADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>870</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2L2_GTGAAA_L002_R1_006.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 374138394</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140605_D00360_0020_AH9V1RADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>871</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2L2_GTGAAA_L002_R1_006.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 483222685</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140528_D00360_0018_AH8VC6ADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>872</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2L2_GTGAAA_L002_R2_001.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 498694039</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140528_D00360_0019_BH8VDAADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>873</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2L2_GTGAAA_L002_R2_001.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 545351336</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140605_D00360_0021_BH9V1VADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>874</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2L2_GTGAAA_L002_R2_001.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 546429577</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140528_D00360_0018_AH8VC6ADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>875</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2L2_GTGAAA_L002_R2_001.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 546928538</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140605_D00360_0020_AH9V1RADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>876</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2L2_GTGAAA_L002_R2_001.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 548835083</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140609_D00360_0022_AH9UJNADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>877</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2L2_GTGAAA_L002_R2_002.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 504197968</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140528_D00360_0019_BH8VDAADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>878</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2L2_GTGAAA_L002_R2_002.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 546476564</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140528_D00360_0018_AH8VC6ADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>879</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2L2_GTGAAA_L002_R2_002.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 547749740</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140605_D00360_0020_AH9V1RADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>880</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2L2_GTGAAA_L002_R2_002.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 549350395</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140609_D00360_0022_AH9UJNADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>881</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2L2_GTGAAA_L002_R2_002.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 552492863</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140605_D00360_0021_BH9V1VADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>882</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2L2_GTGAAA_L002_R2_003.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 497372061</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140528_D00360_0019_BH8VDAADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>883</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2L2_GTGAAA_L002_R2_003.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 546383552</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140528_D00360_0018_AH8VC6ADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>884</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2L2_GTGAAA_L002_R2_003.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 548259636</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140605_D00360_0020_AH9V1RADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>885</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2L2_GTGAAA_L002_R2_003.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 548342149</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140605_D00360_0021_BH9V1VADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>886</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2L2_GTGAAA_L002_R2_003.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 550035823</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140609_D00360_0022_AH9UJNADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>887</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2L2_GTGAAA_L002_R2_004.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 447965207</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140528_D00360_0019_BH8VDAADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>888</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2L2_GTGAAA_L002_R2_004.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 544033430</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140605_D00360_0021_BH9V1VADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>889</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2L2_GTGAAA_L002_R2_004.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 544201960</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140528_D00360_0018_AH8VC6ADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>890</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2L2_GTGAAA_L002_R2_004.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 545543263</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140605_D00360_0020_AH9V1RADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>891</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2L2_GTGAAA_L002_R2_004.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 552307481</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140609_D00360_0022_AH9UJNADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>892</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2L2_GTGAAA_L002_R2_005.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 541975944</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140609_D00360_0022_AH9UJNADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>893</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2L2_GTGAAA_L002_R2_005.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 545293576</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140605_D00360_0020_AH9V1RADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>894</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2L2_GTGAAA_L002_R2_005.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 545590709</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140528_D00360_0018_AH8VC6ADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>895</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2L2_GTGAAA_L002_R2_005.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 548590119</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140605_D00360_0021_BH9V1VADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>896</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2L2_GTGAAA_L002_R2_006.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 188453172</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140609_D00360_0022_AH9UJNADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>897</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2L2_GTGAAA_L002_R2_006.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 366149870</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140605_D00360_0021_BH9V1VADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>898</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2L2_GTGAAA_L002_R2_006.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 399277530</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140605_D00360_0020_AH9V1RADXX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>899</th>\n",
       "      <td> hg002</td>\n",
       "      <td> 2L2_GTGAAA_L002_R2_006.fastq.gz</td>\n",
       "      <td> fastq.gz</td>\n",
       "      <td> 518529864</td>\n",
       "      <td> /home/neffr01/minerva/short_reads/hg002/ftp.nc...</td>\n",
       "      <td> 140528_D00360_0018_AH8VC6ADXX</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>818 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    folder                             name extension       size                                               path                         parent\n",
       "82   hg002  2A1_CGATGT_L001_R1_001.fastq.gz  fastq.gz  428893454  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140528_D00360_0019_BH8VDAADXX\n",
       "83   hg002  2A1_CGATGT_L001_R1_001.fastq.gz  fastq.gz  509951892  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140528_D00360_0018_AH8VC6ADXX\n",
       "84   hg002  2A1_CGATGT_L001_R1_001.fastq.gz  fastq.gz  513979404  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140605_D00360_0021_BH9V1VADXX\n",
       "85   hg002  2A1_CGATGT_L001_R1_001.fastq.gz  fastq.gz  514453620  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140605_D00360_0020_AH9V1RADXX\n",
       "86   hg002  2A1_CGATGT_L001_R1_001.fastq.gz  fastq.gz  523477905  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140609_D00360_0022_AH9UJNADXX\n",
       "87   hg002  2A1_CGATGT_L001_R1_001.fastq.gz  fastq.gz  524462927  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140609_D00360_0023_BH9UD5ADXX\n",
       "88   hg002  2A1_CGATGT_L001_R1_002.fastq.gz  fastq.gz  167071275  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140605_D00360_0020_AH9V1RADXX\n",
       "89   hg002  2A1_CGATGT_L001_R1_002.fastq.gz  fastq.gz  180252467  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140605_D00360_0021_BH9V1VADXX\n",
       "90   hg002  2A1_CGATGT_L001_R1_002.fastq.gz  fastq.gz  203174069  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140528_D00360_0018_AH8VC6ADXX\n",
       "91   hg002  2A1_CGATGT_L001_R1_002.fastq.gz  fastq.gz  206957195  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140609_D00360_0023_BH9UD5ADXX\n",
       "92   hg002  2A1_CGATGT_L001_R1_002.fastq.gz  fastq.gz  225431818  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140609_D00360_0022_AH9UJNADXX\n",
       "93   hg002  2A1_CGATGT_L001_R2_001.fastq.gz  fastq.gz  459671519  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140528_D00360_0019_BH8VDAADXX\n",
       "94   hg002  2A1_CGATGT_L001_R2_001.fastq.gz  fastq.gz  548917388  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140605_D00360_0021_BH9V1VADXX\n",
       "95   hg002  2A1_CGATGT_L001_R2_001.fastq.gz  fastq.gz  549026753  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140528_D00360_0018_AH8VC6ADXX\n",
       "96   hg002  2A1_CGATGT_L001_R2_001.fastq.gz  fastq.gz  554220163  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140605_D00360_0020_AH9V1RADXX\n",
       "97   hg002  2A1_CGATGT_L001_R2_001.fastq.gz  fastq.gz  556144016  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140609_D00360_0023_BH9UD5ADXX\n",
       "98   hg002  2A1_CGATGT_L001_R2_001.fastq.gz  fastq.gz  558736324  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140609_D00360_0022_AH9UJNADXX\n",
       "99   hg002  2A1_CGATGT_L001_R2_002.fastq.gz  fastq.gz  178949495  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140605_D00360_0020_AH9V1RADXX\n",
       "100  hg002  2A1_CGATGT_L001_R2_002.fastq.gz  fastq.gz  192699874  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140605_D00360_0021_BH9V1VADXX\n",
       "101  hg002  2A1_CGATGT_L001_R2_002.fastq.gz  fastq.gz  218221790  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140528_D00360_0018_AH8VC6ADXX\n",
       "102  hg002  2A1_CGATGT_L001_R2_002.fastq.gz  fastq.gz  219959720  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140609_D00360_0023_BH9UD5ADXX\n",
       "103  hg002  2A1_CGATGT_L001_R2_002.fastq.gz  fastq.gz  239419065  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140609_D00360_0022_AH9UJNADXX\n",
       "104  hg002  2A1_CGATGT_L002_R1_001.fastq.gz  fastq.gz  436819051  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140528_D00360_0019_BH8VDAADXX\n",
       "105  hg002  2A1_CGATGT_L002_R1_001.fastq.gz  fastq.gz  509833118  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140528_D00360_0018_AH8VC6ADXX\n",
       "106  hg002  2A1_CGATGT_L002_R1_001.fastq.gz  fastq.gz  514672669  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140605_D00360_0020_AH9V1RADXX\n",
       "107  hg002  2A1_CGATGT_L002_R1_001.fastq.gz  fastq.gz  516033114  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140605_D00360_0021_BH9V1VADXX\n",
       "108  hg002  2A1_CGATGT_L002_R1_001.fastq.gz  fastq.gz  523301733  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140609_D00360_0022_AH9UJNADXX\n",
       "109  hg002  2A1_CGATGT_L002_R1_001.fastq.gz  fastq.gz  526086768  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140609_D00360_0023_BH9UD5ADXX\n",
       "110  hg002  2A1_CGATGT_L002_R1_002.fastq.gz  fastq.gz  172475775  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140605_D00360_0021_BH9V1VADXX\n",
       "111  hg002  2A1_CGATGT_L002_R1_002.fastq.gz  fastq.gz  179401358  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140605_D00360_0020_AH9V1RADXX\n",
       "..     ...                              ...       ...        ...                                                ...                            ...\n",
       "870  hg002  2L2_GTGAAA_L002_R1_006.fastq.gz  fastq.gz  374138394  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140605_D00360_0020_AH9V1RADXX\n",
       "871  hg002  2L2_GTGAAA_L002_R1_006.fastq.gz  fastq.gz  483222685  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140528_D00360_0018_AH8VC6ADXX\n",
       "872  hg002  2L2_GTGAAA_L002_R2_001.fastq.gz  fastq.gz  498694039  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140528_D00360_0019_BH8VDAADXX\n",
       "873  hg002  2L2_GTGAAA_L002_R2_001.fastq.gz  fastq.gz  545351336  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140605_D00360_0021_BH9V1VADXX\n",
       "874  hg002  2L2_GTGAAA_L002_R2_001.fastq.gz  fastq.gz  546429577  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140528_D00360_0018_AH8VC6ADXX\n",
       "875  hg002  2L2_GTGAAA_L002_R2_001.fastq.gz  fastq.gz  546928538  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140605_D00360_0020_AH9V1RADXX\n",
       "876  hg002  2L2_GTGAAA_L002_R2_001.fastq.gz  fastq.gz  548835083  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140609_D00360_0022_AH9UJNADXX\n",
       "877  hg002  2L2_GTGAAA_L002_R2_002.fastq.gz  fastq.gz  504197968  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140528_D00360_0019_BH8VDAADXX\n",
       "878  hg002  2L2_GTGAAA_L002_R2_002.fastq.gz  fastq.gz  546476564  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140528_D00360_0018_AH8VC6ADXX\n",
       "879  hg002  2L2_GTGAAA_L002_R2_002.fastq.gz  fastq.gz  547749740  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140605_D00360_0020_AH9V1RADXX\n",
       "880  hg002  2L2_GTGAAA_L002_R2_002.fastq.gz  fastq.gz  549350395  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140609_D00360_0022_AH9UJNADXX\n",
       "881  hg002  2L2_GTGAAA_L002_R2_002.fastq.gz  fastq.gz  552492863  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140605_D00360_0021_BH9V1VADXX\n",
       "882  hg002  2L2_GTGAAA_L002_R2_003.fastq.gz  fastq.gz  497372061  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140528_D00360_0019_BH8VDAADXX\n",
       "883  hg002  2L2_GTGAAA_L002_R2_003.fastq.gz  fastq.gz  546383552  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140528_D00360_0018_AH8VC6ADXX\n",
       "884  hg002  2L2_GTGAAA_L002_R2_003.fastq.gz  fastq.gz  548259636  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140605_D00360_0020_AH9V1RADXX\n",
       "885  hg002  2L2_GTGAAA_L002_R2_003.fastq.gz  fastq.gz  548342149  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140605_D00360_0021_BH9V1VADXX\n",
       "886  hg002  2L2_GTGAAA_L002_R2_003.fastq.gz  fastq.gz  550035823  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140609_D00360_0022_AH9UJNADXX\n",
       "887  hg002  2L2_GTGAAA_L002_R2_004.fastq.gz  fastq.gz  447965207  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140528_D00360_0019_BH8VDAADXX\n",
       "888  hg002  2L2_GTGAAA_L002_R2_004.fastq.gz  fastq.gz  544033430  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140605_D00360_0021_BH9V1VADXX\n",
       "889  hg002  2L2_GTGAAA_L002_R2_004.fastq.gz  fastq.gz  544201960  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140528_D00360_0018_AH8VC6ADXX\n",
       "890  hg002  2L2_GTGAAA_L002_R2_004.fastq.gz  fastq.gz  545543263  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140605_D00360_0020_AH9V1RADXX\n",
       "891  hg002  2L2_GTGAAA_L002_R2_004.fastq.gz  fastq.gz  552307481  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140609_D00360_0022_AH9UJNADXX\n",
       "892  hg002  2L2_GTGAAA_L002_R2_005.fastq.gz  fastq.gz  541975944  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140609_D00360_0022_AH9UJNADXX\n",
       "893  hg002  2L2_GTGAAA_L002_R2_005.fastq.gz  fastq.gz  545293576  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140605_D00360_0020_AH9V1RADXX\n",
       "894  hg002  2L2_GTGAAA_L002_R2_005.fastq.gz  fastq.gz  545590709  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140528_D00360_0018_AH8VC6ADXX\n",
       "895  hg002  2L2_GTGAAA_L002_R2_005.fastq.gz  fastq.gz  548590119  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140605_D00360_0021_BH9V1VADXX\n",
       "896  hg002  2L2_GTGAAA_L002_R2_006.fastq.gz  fastq.gz  188453172  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140609_D00360_0022_AH9UJNADXX\n",
       "897  hg002  2L2_GTGAAA_L002_R2_006.fastq.gz  fastq.gz  366149870  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140605_D00360_0021_BH9V1VADXX\n",
       "898  hg002  2L2_GTGAAA_L002_R2_006.fastq.gz  fastq.gz  399277530  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140605_D00360_0020_AH9V1RADXX\n",
       "899  hg002  2L2_GTGAAA_L002_R2_006.fastq.gz  fastq.gz  518529864  /home/neffr01/minerva/short_reads/hg002/ftp.nc...  140528_D00360_0018_AH8VC6ADXX\n",
       "\n",
       "[818 rows x 6 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fastq_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "study=\"HG002-GIAB\"\n",
    "left=\"\"\n",
    "right=\"\"\n",
    "sample=\"HG002\"\n",
    "lane=\"\"\n",
    "outfile = open('/home/neffr01/minerva/short_reads.index', 'wb')\n",
    "for i,n in fastq_files.iterrows():\n",
    "    basepath = '/'.join(n['path'].split('/')[0:-1]) + '/'\n",
    "    if \"R1\" not in n['name']:\n",
    "        continue\n",
    "    sample = n['parent'] + \"-\" + \"_\".join(n['name'].split(\"_\")[0:3])\n",
    "    left = n['path']\n",
    "    right = n['path'].replace('_R1_','_R2_')\n",
    "    lane = n['name'].split('_')[2].strip('L')\n",
    "    outfile.write('\\t'.join([left, right, sample, lane, study]) + '\\n')\n",
    "outfile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "bamfiles_quarles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "samples_set = set([\"_\".join(i['name'].split(\"_\")[0:2]) for a,i in bamfiles_quarles.iterrows()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for i in samples_set:\n",
    "    print i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for i in samples_set:\n",
    "    names = bamfiles_quarles[[i == \"_\".join(a.split(\"_\")[0:2]) for a in bamfiles_quarles['name']]]\n",
    "    try:\n",
    "        print \"\\t\".join([list(names[[\"out\" in a for a in names['name']]]['path'])[0],\n",
    "                         list(names[[\"merged_mhgrid.bam_rg.bam\" in a for a in names['name']]]['path'])[0], \n",
    "                        i])\n",
    "    except:\n",
    "        sys.stderr.write(\"error on sample \" + i + \" \\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "sizeof_fmt(sum(pdarr_pretty['size']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "redcap_ids = pd.read_csv('redcap_ids.txt', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "redcap_ids = [i[0] for a,i in redcap_ids.iterrows()]\n",
    "redcap_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "index_list = []\n",
    "for a,i in pdarr_pretty.iterrows():\n",
    "    print i['name']\n",
    "    continue\n",
    "    inlist = any([True for b in redcap_ids if b in str(i['name'])])\n",
    "    if inlist:\n",
    "        index_list.append(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "len(index_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "sizeof_fmt(sum(pdarr_pretty['size']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pdarr_pretty.sort(columns='size', ascending=False, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pdarr_pretty['size'] = [sizeof_fmt(i) for i in pdarr_pretty['size']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pdarr_pretty[(pdarr_pretty['extension']=='fastq.gz')].to_csv('totalRNA.txt', sep='\\t', index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pdarr_pretty[(pdarr_pretty['extension']=='bam')].to_csv('/data/projects_gibbons/home/neffra/mhgrid_tot_rna_bams', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "sorted(filedict.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "res = pdarr_pretty[pdarr_pretty.duplicated(cols='name') | pdarr_pretty.duplicated(cols='name', take_last=True)].sort(columns='name')\n",
    "res2 = res[res.duplicated(cols='size') | res.duplicated(cols='size', take_last=True)].sort(columns='name')\n",
    "res2['path'] = ['/'.join(i.split('/')[6:]) for i in list(res2['path'])]\n",
    "res3 = res2[[('QC' not in i) & ('.txt' not in i) for i in res2['path']]]\n",
    "\n",
    "duplevel = 0\n",
    "for i in res3.iterrows():\n",
    "    duplevel += int(i[1][3])\n",
    "\n",
    "print \"Size of duplicated files: \" + sizeof_fmt(duplevel/2)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#use this to look for files within the directory\n",
    "\n",
    "ext = \"vcf\"\n",
    "totsize = 0\n",
    "for i in list(pdarr_pretty[[ext in i for i in pdarr_pretty['extension']]]['size']):\n",
    "    totsize += int(i)\n",
    "\n",
    "print \"\\nTotal size: \" + sizeof_fmt(totsize)\n",
    "                                    \n",
    "display.display_html(pdarr_pretty[[ext in i for i in pdarr_pretty['extension']]])\n",
    "\n",
    "a = pdarr_pretty[[ext in i for i in pdarr_pretty['extension']]]\n",
    "a.to_excel('vcffiles.xlsx', 'VCFFiles')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#use this to look for files within the directory\n",
    "\n",
    "ext = \"fastq.gz\"\n",
    "totsize = 0\n",
    "for i in list(pdarr_pretty[[ext in i for i in pdarr_pretty['extension']]]['size']):\n",
    "    totsize += int(i)\n",
    "\n",
    "print \"\\nTotal uncompressed size: \" + sizeof_fmt(totsize)\n",
    "\n",
    "g = pdarr_pretty[[ext in i for i in pdarr_pretty['extension']]]\n",
    "g['name'] = ['_'.join(x.split(\"_\")[0:4]) for x in g['name']]\n",
    "count = 0\n",
    "\n",
    "totsize = 0\n",
    "#merge_file = open(\"/data/test/home/neffra/mhgrid_analysis/scripts/merge_all_mhgrid.sh\", \"wb\")\n",
    "alt_file = open(\"/data/test/home/neffra/mhgrid_analysis/align_files2\", \"wb\")\n",
    "for name, group in pdarr_pretty[[\"MHM\" in i for i in pdarr_pretty['name']]].groupby(g['name']):\n",
    "    #names = sorted(list(set('_'.join(x.split(\"_\")[0:4]) for x in group['name'])))\n",
    "    #names = [i + \"_bwamem.realigned.bam\" for i in names]\n",
    "    #merge_file.write('qsub -b y -N merge -o /dev/null -e mergeout.err \"samtools merge ' + name + \"_merged_mhgrid.bam \" + \" \".join(names) + '\"\\n')\n",
    "    alt_file.write(\"_\".join(str(list(group['path'])[0]).split(\"_\")[0:-2]) + \" \" + name + \"\\n\")\n",
    "    #l = str(list(group['path'])[0]).split(\"/\")[0:-1]\n",
    "    #l.append(name)\n",
    "    #print '/'.join(l)\n",
    "    #totsize += np.sum(group['size'])\n",
    "#    print len(group)\n",
    "    #count += 1\n",
    "    #if count > 50: \n",
    "     #   break\n",
    "alt_file.close()\n",
    "#merge_file.close()\n",
    "print sizeof_fmt(totsize/count)\n",
    "print count\n",
    "                                    \n",
    "print list(g[[\"MHM\" in i for i in g['name']]].head(1)['path'])\n",
    "\n",
    "#for i in list(pdarr_pretty[[ext in i for i in pdarr_pretty['extension']]]['path'])[0:10]:\n",
    "#    html = '<a href=\"sftp://neffra@walnut.nhgri.nih.gov'\n",
    "#    html += i\n",
    "#    html += '\">' + i.split('/')[-1] + \"</a>\"\n",
    "#    h = HTML(html)\n",
    "#    display.display_html(h)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "ext = \"fastq.gz\"\n",
    "study = \"S1304120MHM\"\n",
    "totsize = 0\n",
    "for i in list(pdarr_pretty[[ext in i for i in pdarr_pretty['extension']]]['size']):\n",
    "    totsize += int(i)\n",
    "\n",
    "print \"\\nTotal uncompressed size: \" + sizeof_fmt(totsize)\n",
    "\n",
    "f = pdarr_pretty[[study in i for i in pdarr_pretty['name']]]\n",
    "g = f[[ext in i for i in f['extension']]]\n",
    "g['name'] = ['_'.join(x.split(\"_\")[0:2]) for x in g['name']]\n",
    "\n",
    "print len(set(g['name']))\n",
    "g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "masterdict = dict()\n",
    "for name, group in grouped:\n",
    "    print name\n",
    "    filedict = dict()\n",
    "    \n",
    "    for idx, item in group.iterrows():\n",
    "        ext = item[2]\n",
    "        if ext in filedict:\n",
    "            filedict[ext][0] += item[3]\n",
    "            filedict[ext][1] += 1\n",
    "        else:\n",
    "            filedict[ext] = [item[3], 1]\n",
    "    tbl = ListTable()\n",
    "    tbl.append(['extension', '# files', 'size'])\n",
    "    for key in sorted(filedict.keys()):\n",
    "        tbl.append([key, str(filedict[key][1]), sizeof_fmt(filedict[key][0])])\n",
    "    display.display_html(tbl)\n",
    "    for key in filedict:\n",
    "        if key in masterdict:\n",
    "            masterdict[key][0] += filedict[key][0]\n",
    "            masterdict[key][1] += filedict[key][1]\n",
    "        else:\n",
    "            masterdict[key] = filedict[key]\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "print \"---------------\\n\"\n",
    "print \"Master stats:\\n\"\n",
    "totsize = 0\n",
    "tbl = ListTable()\n",
    "tbl.append(['extension', '# files', 'size'])\n",
    "for key in sorted(masterdict.keys()):\n",
    "    if masterdict[key][0] >= 1*1024*1024*512:\n",
    "        tbl.append([key, str(masterdict[key][1]), sizeof_fmt(masterdict[key][0])])\n",
    "    totsize += masterdict[key][0]\n",
    "print sizeof_fmt(totsize)\n",
    "display.display_html(tbl)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
